{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PB Deletions\n",
    "\n",
    "* Deletions\n",
    "* HG002 Only\n",
    "* Technology: PB\n",
    "\n",
    "* The following classifier will train on and predict the following labels:\n",
    "    - GTcons\n",
    "    \n",
    "    \n",
    "* All -1 values have been removed from the train and test set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/sklearn/cross_validation.py:44: DeprecationWarning:\n",
      "\n",
      "This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.\n",
      "\n",
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/sklearn/grid_search.py:43: DeprecationWarning:\n",
      "\n",
      "This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Imports\n",
    "\"\"\"\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from fancyimpute import KNN\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.model_selection import LeaveOneOut\n",
    "from scipy.stats import ks_2samp\n",
    "from scipy import stats\n",
    "from matplotlib import pyplot\n",
    "from sklearn import preprocessing\n",
    "from scipy.linalg import svd\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import roc_auc_score\n",
    "import seaborn as sns\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.decomposition import PCA as sklearnPCA\n",
    "import plotly.plotly as py\n",
    "from sklearn.cluster import DBSCAN\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import metrics\n",
    "from sklearn.grid_search import GridSearchCV\n",
    "from sklearn.metrics import f1_score, precision_score\n",
    "from sklearn import preprocessing\n",
    "from ggplot import *\n",
    "from bokeh.charts import TimeSeries\n",
    "from bokeh.models import HoverTool\n",
    "from bokeh.plotting import show\n",
    "from bokeh.charts import Scatter, Histogram, output_file, show\n",
    "from bokeh.plotting import figure, show, output_file, ColumnDataSource\n",
    "from bokeh.io import output_notebook\n",
    "from bokeh.charts import Bar, output_file, show\n",
    "import bokeh.palettes as palettes\n",
    "from bokeh.models import HoverTool, BoxSelectTool, Legend\n",
    "from sklearn import (manifold, datasets, decomposition, ensemble,\n",
    "                     discriminant_analysis, random_projection)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Import Data\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>id</th>\n",
       "      <th>sample</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>SVtype</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>39</td>\n",
       "      <td>HG002</td>\n",
       "      <td>109690878</td>\n",
       "      <td>109690918</td>\n",
       "      <td>Insertion</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>-39</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>222.0</td>\n",
       "      <td>197.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>20597.94118</td>\n",
       "      <td>5844.767921</td>\n",
       "      <td>17.0</td>\n",
       "      <td>12582.47059</td>\n",
       "      <td>3520.607336</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  id sample      start        end       type    SVtype  Size  refN_cnt  \\\n",
       "0     1  39  HG002  109690878  109690918  Insertion  Deletion   -39         0   \n",
       "\n",
       "   refN_pct                ...                 \\\n",
       "0         0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            222.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       197.0                            11.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                            9.0               20597.94118   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              5844.767921              17.0                 12582.47059   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                3520.607336                              17.0  \n",
       "\n",
       "[1 rows x 40 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import Training Data\n",
    "# SVanalyzer generated training data\n",
    "df_train = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Sept122017/Step3_ML/data/train/tech_sep/DEL/no_minus_one/pacbio_HG002_no-1_DEL.csv')\n",
    "df_train_2 = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Sept122017/Step3_ML/data/train/tech_sep/DEL/no_minus_one/pacbio_HG002_no-1_DEL.csv')\n",
    "df_train.rename(columns={'size': 'Size'}, inplace=True)\n",
    "df_train.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "train_set = pd.DataFrame()\n",
    "train_set = df_train_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "train_set['GTcons'].replace(0, 'Homozygous_Reference', inplace=True)\n",
    "train_set['GTcons'].replace(1, 'Heterozygous_Variant', inplace=True)\n",
    "train_set['GTcons'].replace(2, 'Homozygous_Variant', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Homozygous_Reference    1176\n",
       "Heterozygous_Variant     642\n",
       "Homozygous_Variant       154\n",
       "dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(train_set['GTcons'].values, sort=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:**\n",
    "    The original training set has 154 homozygous variant examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Train the model only on the rows that have an Exact Match or Homozygous Reference Label\n",
    "# This step removes any row that has in 'Inaccurate Call' label\n",
    "df_train = df_train[(df_train['Label'] == 1) | (df_train['Label'] == 0)]\n",
    "df_train_2 = df_train_2[(df_train_2['Label'] == 1) | (df_train_2['Label'] == 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    685\n",
       "1    535\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(df_train['Label'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    685\n",
       "1    535\n",
       "dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#(Copy Training Set) Train only on the rows that have an Exact Match or Homozygous Reference Label\n",
    "# The following lists the count of EM, HR, Inaccurate Match Labels\n",
    "pd.value_counts(df_train_2['Label'].values, sort=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='hom_ref'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Homozygous_Reference    669\n",
       "Heterozygous_Variant    427\n",
       "Homozygous_Variant      124\n",
       "dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List the consensus GT count (generated from R script) within the training set. Reminder! This dataset only includes svanalyzer Exact Match and Homozygous Reference datapoints only\n",
    "pd.value_counts(df_train_2['GTcons'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>id</th>\n",
       "      <th>sample</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>type</th>\n",
       "      <th>SVtype</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>HG002</td>\n",
       "      <td>95498964</td>\n",
       "      <td>95499280</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>-316</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>215.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  id sample     start       end      type    SVtype  Size  refN_cnt  \\\n",
       "0     4   1  HG002  95498964  95499280  Deletion  Deletion  -316         0   \n",
       "\n",
       "   refN_pct                ...                 \\\n",
       "0         0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            215.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       255.0                             0.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           12.0                       0.0   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0                      0.0               0.0                         0.0   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                        0.0                               0.0  \n",
       "\n",
       "[1 rows x 39 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import Test Data\n",
    "# SVanalyzer generated training data\n",
    "df_test = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Sept122017/Step3_ML/data/test/tech_sep/DEL/no_minus_one/pacbio_HG002_no-1_DEL.csv')\n",
    "df_test_2 = pd.read_csv('/Volumes/lesleydata/SVanalyzer_ML/Sept122017/Step3_ML/data/test/tech_sep/DEL/no_minus_one/pacbio_HG002_no-1_DEL.csv')\n",
    "df_test.rename(columns={'size': 'Size'}, inplace=True)\n",
    "df_test.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Label'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store header names in lists and find names that are NOT contained in BOTH lists\n",
    "c = list(df_train.columns.values)\n",
    "d = list(df_test.columns.values)\n",
    "set(c) - set(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Drop columns that are not shared by both dataframes\n",
    "df_train.drop(['Label'], axis=1, inplace = True)\n",
    "df_train.drop(['GTconflict'], axis=1, inplace = True)\n",
    "df_train.drop(['GTsupp'], axis=1, inplace = True)\n",
    "df_train.drop(['sample'], axis=1, inplace = True)\n",
    "df_train.drop(['SVtype'], axis=1, inplace = True)\n",
    "df_train.drop(['type'], axis=1, inplace = True)\n",
    "df_train.drop(['id'], axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>109690878</td>\n",
       "      <td>109690918</td>\n",
       "      <td>-39</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.4</td>\n",
       "      <td>...</td>\n",
       "      <td>222.0</td>\n",
       "      <td>197.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>20597.94118</td>\n",
       "      <td>5844.767921</td>\n",
       "      <td>17.0</td>\n",
       "      <td>12582.47059</td>\n",
       "      <td>3520.607336</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom      start        end  Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0     1  109690878  109690918   -39         0         0           0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0         0.0              1            0.4                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            222.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       197.0                            11.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                            9.0               20597.94118   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              5844.767921              17.0                 12582.47059   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                3520.607336                              17.0  \n",
       "\n",
       "[1 rows x 33 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_train['chrom'].replace('X', 23, inplace=True)\n",
    "df_train['chrom'].replace('Y', 24, inplace=True)\n",
    "df_test['chrom'].replace('X', 23, inplace=True)\n",
    "df_test['chrom'].replace('Y', 24, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GTconflict', 'GTsupp', 'SVtype', 'id', 'sample', 'type'}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store header names in lists and find names that are NOT contained in BOTH lists\n",
    "c = list(df_train.columns.values)\n",
    "d = list(df_test.columns.values)\n",
    "set(d) - set(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Drop columns that are not shared by both dataframes\n",
    "df_test.drop(['GTcons'], axis=1, inplace = True)\n",
    "df_test.drop(['GTconflict'], axis=1, inplace = True)\n",
    "df_test.drop(['GTsupp'], axis=1, inplace = True)\n",
    "df_test.drop(['sample'], axis=1, inplace = True)\n",
    "df_test.drop(['SVtype'], axis=1, inplace = True)\n",
    "df_test.drop(['type'], axis=1, inplace = True)\n",
    "df_test.drop(['id'], axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Impute missing values using KNN\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>109690878</td>\n",
       "      <td>109690918</td>\n",
       "      <td>-39</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0.4</td>\n",
       "      <td>...</td>\n",
       "      <td>222.0</td>\n",
       "      <td>197.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>20597.94118</td>\n",
       "      <td>5844.767921</td>\n",
       "      <td>17.0</td>\n",
       "      <td>12582.47059</td>\n",
       "      <td>3520.607336</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1869280</td>\n",
       "      <td>1869312</td>\n",
       "      <td>-31</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>177.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>16622.10000</td>\n",
       "      <td>6461.411834</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10433.70000</td>\n",
       "      <td>3505.460085</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2891239</td>\n",
       "      <td>2891284</td>\n",
       "      <td>-44</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>179.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom      start        end  Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0     1  109690878  109690918   -39         0         0           0   \n",
       "1     1    1869280    1869312   -31         0         0           0   \n",
       "2     1    2891239    2891284   -44         0         0           1   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0    0.000000              1            0.4                ...                  \n",
       "1    0.000000              0            0.0                ...                  \n",
       "2    0.666667              1            1.0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            222.0   \n",
       "1                                            177.0   \n",
       "2                                            179.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       197.0                            11.0   \n",
       "1                       125.0                             5.0   \n",
       "2                       201.0                             1.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                            9.0               20597.94118   \n",
       "1                            6.0               16622.10000   \n",
       "2                           13.0                   0.00000   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              5844.767921              17.0                 12582.47059   \n",
       "1              6461.411834              10.0                 10433.70000   \n",
       "2                 0.000000               0.0                     0.00000   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                3520.607336                              17.0  \n",
       "1                3505.460085                              10.0  \n",
       "2                   0.000000                               0.0  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store training data in a new variable which will be converted to a matrix\n",
    "X = df_train\n",
    "X.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imputing row 1/1220 with 0 missing, elapsed time: 0.398\n",
      "Imputing row 101/1220 with 0 missing, elapsed time: 0.399\n",
      "Imputing row 201/1220 with 0 missing, elapsed time: 0.399\n",
      "Imputing row 301/1220 with 0 missing, elapsed time: 0.400\n",
      "Imputing row 401/1220 with 0 missing, elapsed time: 0.400\n",
      "Imputing row 501/1220 with 0 missing, elapsed time: 0.401\n",
      "Imputing row 601/1220 with 0 missing, elapsed time: 0.401\n",
      "Imputing row 701/1220 with 0 missing, elapsed time: 0.402\n",
      "Imputing row 801/1220 with 0 missing, elapsed time: 0.451\n",
      "Imputing row 901/1220 with 0 missing, elapsed time: 0.451\n",
      "Imputing row 1001/1220 with 0 missing, elapsed time: 0.452\n",
      "Imputing row 1101/1220 with 0 missing, elapsed time: 0.452\n",
      "Imputing row 1201/1220 with 0 missing, elapsed time: 0.453\n"
     ]
    }
   ],
   "source": [
    "# Convert dataframe to matrix\n",
    "X=X.as_matrix()\n",
    "\n",
    "#Imput missing values from three closest observations\n",
    "X_imputed=KNN(k=3).complete(X)\n",
    "X=pd.DataFrame(X_imputed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>109690878.0</td>\n",
       "      <td>109690918.0</td>\n",
       "      <td>-39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>...</td>\n",
       "      <td>222.0</td>\n",
       "      <td>197.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>20597.94118</td>\n",
       "      <td>5844.767921</td>\n",
       "      <td>17.0</td>\n",
       "      <td>12582.47059</td>\n",
       "      <td>3520.607336</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1869280.0</td>\n",
       "      <td>1869312.0</td>\n",
       "      <td>-31.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>177.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>16622.10000</td>\n",
       "      <td>6461.411834</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10433.70000</td>\n",
       "      <td>3505.460085</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2891239.0</td>\n",
       "      <td>2891284.0</td>\n",
       "      <td>-44.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>179.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom        start          end  Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0    1.0  109690878.0  109690918.0 -39.0       0.0       0.0         0.0   \n",
       "1    1.0    1869280.0    1869312.0 -31.0       0.0       0.0         0.0   \n",
       "2    1.0    2891239.0    2891284.0 -44.0       0.0       0.0         1.0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0    0.000000            1.0            0.4                ...                  \n",
       "1    0.000000            0.0            0.0                ...                  \n",
       "2    0.666667            1.0            1.0                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            222.0   \n",
       "1                                            177.0   \n",
       "2                                            179.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       197.0                            11.0   \n",
       "1                       125.0                             5.0   \n",
       "2                       201.0                             1.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                            9.0               20597.94118   \n",
       "1                            6.0               16622.10000   \n",
       "2                           13.0                   0.00000   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0              5844.767921              17.0                 12582.47059   \n",
       "1              6461.411834              10.0                 10433.70000   \n",
       "2                 0.000000               0.0                     0.00000   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                3520.607336                              17.0  \n",
       "1                3505.460085                              10.0  \n",
       "2                   0.000000                               0.0  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Store header values in a list, will be used later to re-label the matrix post KNN imputation\n",
    "dftrain_header = list(df_train.columns.values)\n",
    "X.columns = dftrain_header\n",
    "X.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Store Labels in a new 'Y' DataFrame\n",
    "Y = pd.DataFrame()\n",
    "Y['GTcons'] = X['GTcons']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0    427\n",
       "2.0    124\n",
       "0.0    669\n",
       "dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Count the number of labels\n",
    "pd.value_counts(Y['GTcons'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Note: originally selected 1000 of each label --> find out why some are lost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Remove labels from feature set\n",
    "X.drop(['GTcons'],axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Order features\n",
    "X4 = X.reindex_axis(sorted(X.columns), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Machine Learning\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='machine_learning'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description:\n",
    "\n",
    "   * In the following section a random forest model will be trained on svanalyzer data.\n",
    "\n",
    "       * The model was trained using [train/test split](http://scikit-learn.org/0.16/modules/generated/sklearn.cross_validation.train_test_split.html) where 70% of the data was used to train the model and the model performance was determined by predicting labels for the remaining 30% of the data. The trained model will be used in a [later section](#predict) to predict the consensus GT for 5000 randomly selected deletions [these deletions were randomly selected from [union_170509_refalt.sort.vcf](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/NIST_UnionSVs_05092017/)]\n",
    "       * In the following section, svanalyzer data was used to train a random forest (RF) model. The features for the svanalyzer dataset include: svviz features, GA4GH features [RefN, Segmental Duplications, Tandem Repeat], preliminary R script analysis [consensus GT, GTsup].\n",
    "       * The RF classifier will predict the consensus GT labels:\n",
    "           * Homozygous Reference (0)\n",
    "           * Heterozygous Variant (1)\n",
    "           * Homozygous Variant (2)\n",
    "       \n",
    "       * In the [following section](#prediction_step), the trained RF model will be used to predict labels for genotype labels for 5000 randomly selected deletions [these deletions were randomly selected from [union_170509_refalt.sort.vcf](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/NIST_UnionSVs_05092017/)]. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Train Random Forest Classifier **"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='train_test'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Train Test Split\n",
    "# Train on 70% of the data and test on 30%\n",
    "X_train, X_test, y_train, y_test = train_test_split(X4, Y, test_size=0.3, random_state=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/ipykernel/__main__.py:4: DataConversionWarning:\n",
      "\n",
      "A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',\n",
       "            max_depth=None, max_features='auto', max_leaf_nodes=None,\n",
       "            min_impurity_split=1e-07, min_samples_leaf=1,\n",
       "            min_samples_split=2, min_weight_fraction_leaf=0.0,\n",
       "            n_estimators=10, n_jobs=1, oob_score=False, random_state=None,\n",
       "            verbose=0, warm_start=False)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %%timeit\n",
    "# Train Random Forest Classifier\n",
    "model = RandomForestClassifier() \n",
    "model.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Feature ranking:\n",
      "1. feature 5 (0.290156)\n",
      "2. feature 8 (0.210269)\n",
      "3. feature 9 (0.080270)\n",
      "4. feature 7 (0.071467)\n",
      "5. feature 3 (0.043643)\n",
      "6. feature 6 (0.023884)\n",
      "7. feature 20 (0.023385)\n",
      "8. feature 12 (0.019432)\n",
      "9. feature 19 (0.019049)\n",
      "10. feature 11 (0.017917)\n",
      "11. feature 22 (0.016557)\n",
      "12. feature 16 (0.016073)\n",
      "13. feature 2 (0.015490)\n",
      "14. feature 24 (0.015482)\n",
      "15. feature 13 (0.014365)\n",
      "16. feature 10 (0.013881)\n",
      "17. feature 29 (0.011915)\n",
      "18. feature 23 (0.011462)\n",
      "19. feature 0 (0.011331)\n",
      "20. feature 4 (0.010753)\n",
      "21. feature 18 (0.010713)\n",
      "22. feature 14 (0.010482)\n",
      "23. feature 15 (0.010133)\n",
      "24. feature 21 (0.009354)\n",
      "25. feature 1 (0.008740)\n",
      "26. feature 17 (0.007289)\n",
      "27. feature 28 (0.002213)\n",
      "28. feature 30 (0.002135)\n",
      "29. feature 31 (0.001407)\n",
      "30. feature 27 (0.000750)\n",
      "31. feature 25 (0.000000)\n",
      "32. feature 26 (0.000000)\n"
     ]
    }
   ],
   "source": [
    "# Feature Importance\n",
    "importances = model.feature_importances_\n",
    "indices = np.argsort(importances)[::-1]\n",
    "print(\"Feature ranking:\")\n",
    "\n",
    "for f in range(X_test.shape[1]):\n",
    "    print(\"%d. feature %d (%f)\" % (f + 1, indices[f], importances[indices[f]]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Another Resource](https://stackoverflow.com/questions/37877542/how-to-label-the-feature-importance-with-forests-of-trees)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/lmc2/anaconda/envs/NIHFAES/lib/python3.5/site-packages/ipykernel/__main__.py:3: FutureWarning:\n",
      "\n",
      "sort is deprecated, use sort_values(inplace=True) for INPLACE sorting\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x116027e10>"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAHeCAYAAABg00vYAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsnXeYZFWZ/z8zDEFgQMKIoi6uSn8FRQyIIIiii7qKghFl\nWf2BqBgQBROIAbPuYlpBgiDJgIiIAQmKCRAxrYDKiwgiCJKcHZIgw/Tvj/dW9+3qG865VTXT1Lyf\n5+mnu6tOnXvPPafee+4b501OThIEQRCMF/NX9AkEQRAEwyeEexAEwRgSwj0IgmAMCeEeBEEwhoRw\nD4IgGENCuAdBEIwhC9oaSJoPHA5sCdwN7G1mV5TefzHwLmAS+JKZfabtM1XcdNNts3wy11tvTRYv\nvjNpIKlth90u+lw5+xy38USf9931sWjRwnlV7VN27rsCa5jZtrgQP7T3hqRVgI8B/wZsC7xB0oZN\nn8lhwYJVht522O2iz5Wzz3EbT/Q5XusD0oT79sCZAGZ2IbBV7w0zuxfYzMyWABsAqwD/bPpMEARB\nMHrmtUWoSvoCcKqZfa/4/y/Aw81saanNi4DDgO8CrwOObPtMP0uX3juZe2cKgiAIqFTLtOrcgVuB\nhaX/5/cLaTP7hqRvAscBr0z5TD81uiRuuum2hFNMbzvsdtHnytnnuI0n+rzvro9FixZWtE5Ty5wP\nPBdA0jbAJb03JK0j6ceSVjezZcAdwLKmzwRBEASjJ2Xnfhqwk6QL8O3/npJ2B9Y2s6MkfQn4iaR7\ngIuBk3DPmRmfGc3pB0EQBFW0CvdiR75P38uXld4/Cjiq4qP9nwmCIAiWExHEFARBMIaEcA+CIBhD\nUnTuQRAEwQpir4+dO+u1Y9/1jNbPxc49CIJgDAnhHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ\n7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5B\nEARjSAj3IAiCMSSEexAEwRgSwj0IgmAMCeEeBEEwhoRwD4IgGENCuAdBEIwhIdyDIAjGkBDuQRAE\nY0gI9yAIgjEkhHsQBMEYEsI9CIJgDFnQ1kDSfOBwYEvgbmBvM7ui9P4rgLcAS4FLgDeY2TJJvwZu\nLZpdZWZ7DvvkgyAIgmpahTuwK7CGmW0raRvgUGAXAEn3Az4EbGFmd0r6CrCzpLOBeWb29BGddxAE\nQdBAilpme+BMADO7ENiq9N7dwFPM7M7i/wXAXfguf01JZ0s6t7gpBEEQBMuJeZOTk40NJH0BONXM\nvlf8/xfg4Wa2tK/dvsBzi5/HANsAXwA2Bb4HqP8zZZYuvXdywYJVBhhKEATB+PH8A06f9dq3D92l\n/O+8qs+lqGVuBRaW/p9fFtKFTv4TwATwYjOblHQ5cIWZTQKXS7oFeBBwTd1BFi++c9ZrixYt5Kab\nbks4xfS2w24Xfa6cfY7beKLPub8+ypTbL1q0sLJNilrmfHw3TqFeuaTv/SOBNYBdS+qZvXDdPJI2\nBtYBrk8/9SAIgmAQUnbupwE7SboA3/7vKWl3YG3gl8CrgZ8C50oC+AxwDHCcpPOASWCvJpVMEARB\nMFxahbuZLQP26Xv5stLfdbv/3bueVBAEQTAYEcQUBEEwhoRwD4IgGENCuAdBEIwhIdyDIAjGkBDu\nQRAEY0gI9yAIgjEkhHsQBMEYEsI9CIJgDAnhHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQ\nBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARj\nSAj3IAiCMSSEexAEwRiyYEWfQBAEwcrGXh87t/L1Y9/1jKEdI3buQRAEY0gI9yAIgjEkhHsQBMEY\nEsI9CIJgDGk1qEqaDxwObAncDextZleU3n8F8BZgKXAJ8IbirdrPBEEQBKMlZee+K7CGmW0LvAs4\ntPeGpPsBHwJ2NLPtgHWBnZs+EwRBEIyeFOG+PXAmgJldCGxVeu9u4Clmdmfx/wLgrpbPBEEQBCMm\nxc99HWBJ6f97JS0ws6Vmtgy4AUDSvsDawDnAy+o+U3eQ9dZbkwULVpn1+qJFCxNOMa/tsNtFnytn\nn+M2nuhzxa2PUfSdItxvBco9zS8L6UIn/wlgAnixmU1KavxMFYsX3znrtUWLFnLTTbclnGJ622G3\niz5Xzj7HbTzR54pbH2VSP1NuVyfoU9Qy5wPPBZC0DW40LXMksAawa0k90/aZIAiCYISk7NxPA3aS\ndAEwD9hT0u64CuaXwKuBnwLnSgL4TNVnRnDuQRAEQQ2twr3Qq+/T9/Jlpb/rdv/9nwmCIAiWExHE\nFARBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ4R4E\nQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgSwj0IgmAMCeEeBEEw\nhoRwD4IgGENCuAdBEIwhIdyDIAjGkBDuQRAEY0gI9yAIgjEkhHsQBMEYEsI9CIJgDAnhHgRBMIaE\ncA+CIBhDQrgHQRCMIQvaGkiaDxwObAncDextZlf0tVkTOAd4tZldVrz2a+DWoslVZrbnME88CIIg\nqKdVuAO7AmuY2baStgEOBXbpvSlpK+AI4CGl19YA5pnZ04d7ukEQBEEKKWqZ7YEzAczsQmCrvvdX\nB14IXFZ6bUtgTUlnSzq3uCkEQRAEy4l5k5OTjQ0kfQE41cy+V/z/F+DhZra0r92PgH3M7DJJWwDb\nAF8ANgW+B6j/M2WWLr13csGCVQYZSxAEwX2C5x9weuXr3z50l1mvVbXtazevqq8UtcytwMLS//Ob\nhHTB5cAVZjYJXC7pFuBBwDV1H1i8+M5Zry1atJCbbrot4RTT2w67XfS5cvY5buOJPlfc+iiT+ply\nu0WLFla2SVHLnA88F6BQr1yS8Jm9cN08kjYG1gGuT/hcEARBMARSdu6nATtJugDf/u8paXdgbTM7\nquYzxwDHSToPmAT2StjtB0EQBEOiVbib2TJgn76XL6to9/TS3/8Edh/05IIgCIJuRBBTEATBGBLC\nPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQjCEh3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgSwj0I\ngmAMCeEeBEEwhoRwD4IgGENCuAdBEIwhIdyDIAjGkBDuQRAEY0gI9yAIgjEkhHsQBMEYEsI9CIJg\nDAnhHgRBMIaEcA+CIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ\n4R4EQTCGhHAPgiAYQxa0NZA0Hzgc2BK4G9jbzK7oa7MmcA7wajO7LOUzQRAEwehI2bnvCqxhZtsC\n7wIOLb8paSvgJ8AjUj8TBEEQjJYU4b49cCaAmV0IbNX3/urAC4HLMj4TBEEQjJBWtQywDrCk9P+9\nkhaY2VIAMzsfQFLyZ6pYb701WbBglVmvL1q0MOEU89oOu130uXL2OW7jiT5X3PoYRd8pwv1WoNzT\n/CYh3fUzixffOeu1RYsWctNNtyWcYnrbYbeLPlfOPsdtPNHnilsfZVI/U25XJ+hT1DLnA88FkLQN\ncMmIPhMEQRAMiZSd+2nATpIuAOYBe0raHVjbzI5K/cxQzjYIgiBIolW4m9kyYJ++ly+raPf0ls8E\nQRAEy4kIYgqCIBhDQrgHQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxJIR7EATBGBLCPQiCYAwJ4R4E\nQTCGhHAPgiAYQ1LSDwRBEAQJ7PWxc2e9duy7nrECziR27kEQBGNJCPcgCIIxJIR7EATBGBLCPQiC\nYAwJ4R4EQTCGhHAPgiAYQ8IVMgiCoIEq90ZYcS6OqcTOPQiCYAwJ4R4EQTCGhHAPgiAYQ0K4B0EQ\njCEh3IMgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgSwj0IgmAMCeEeBEEwhoRwD4IgGENac8tImg8c\nDmwJ3A3sbWZXlN5/PvBeYClwrJkdXbz+a+DWotlVZrbnkM89CIIgqCElcdiuwBpmtq2kbYBDgV0A\nJK0KfAp4EnAHcL6kbwFLgHlm9vSRnHUQBEHQSIpaZnvgTAAzuxDYqvTeZsAVZrbYzP4JnAfsgO/y\n15R0tqRzi5tCEARBsJxI2bmvg+/Ee9wraYGZLa147zZgXeBO4L+BLwCbAt+TpOIzlay33posWLDK\nrNcXLVqYcIp5bYfdLvpcOfsct/FEn+ltV/Q5pLRLEe63AuWe5peEdP97C4H/Ay7Hd/STwOWSbgEe\nBFxTd5DFi++c9dqiRQu56abbEk4xve2w20WfK2ef4zae6DO9bY/U9jn9dumzTtCnqGXOB54LUKhX\nLim99wdgU0nrS1oNV8n8DNgL180jaWN8h3990lkHQRAEA5Oycz8N2EnSBcA8YE9JuwNrm9lRkvYH\nzsJvFMea2V8lHQMcJ+k8YBLYq0klEwRBEAyXVuFuZsuAffpevqz0/reBb/d95p/A7sM4wSAIglFQ\nVT5vrpfOyyGCmIIgCMaQEO5BEARjSAj3IAiCMSSEexAEwRgSwj0IgmAMCeEeBEEwhoRwD4IgGENC\nuAdBEIwhIdyDIAjGkBDuQRAEY0gI9yAIgjEkhHsQBMEYEsI9CIJgDAnhHgRBMIak5HMPgiBYoaSm\n561qV9d23ImdexAEwRgSwj0IgmAMCbVMEARDJafC0bhXQ1qRhHAPgiCJEMT3LUK4B8F9hFHsiENg\njy8h3INgBOR4bYSADUZBCPcgyCAEcXBfIYR7cJ9iFDvi8I0OxpEQ7sEUo9DThu43CFYMIdzvo4TQ\nDIKgiRDuc4gQxEEQDIuIUA2CIBhDWnfukuYDhwNbAncDe5vZFaX3nw+8F1gKHGtmR7d9JgiCIBgt\nKTv3XYE1zGxb4F3Aob03JK0KfAp4FvA04LWSNmr6TBAEQTB6UnTu2wNnApjZhZK2Kr23GXCFmS0G\nkHQesAOwbcNnxoIIUgmCYC6TItzXAZaU/r9X0gIzW1rx3m3Aui2fWe6EIA6CYGVj3uTkZGMDSZ8E\nLjSzrxX/X2tmDyn+fizwMTN7bvH/p4DzgafUfSYIgiAYPSk69/OBnvDeBrik9N4fgE0lrS9pNVwl\n87OWzwRBEAQjJmXn3vN8eSwwD9gTeAKwtpkdVfKWmY97yxxW9Rkzu2x0wwiCIAjKtAr3IAiC4L5H\nBDEFQRCMISHcgyAIxpAQ7kEQBGNICPcgCIIxJIT7mCPpIX3/a0WdSxCMmljv08y5lL+SPmdmbyr9\nf4KZvbKvzSrAKsBXgd1wd8v5wBlm9ozcdqM4dk6fpff+pe+le4CbzeyeinN4NLBG7zUzu6ivzWOA\nBwMfl/SO4uVVgI8Cj6s49oOBjwMPAE4BLjazn1edZw6S1gGWAS8EvtNLVVHRbh7wpL4x/aSi3WOA\nzwPrAScBl5rZd/ra7FB3PjV9HmxmHyr9/1EzO7BlaI1IeijwCmaO5wM1bVcB/h+wCXAuPqabBzh2\n0joq2u4E7A+sXjrP8nfovXXHaRhP41zmfIck/RCodOnrO8/k9S7pZDPbrW5cFeNpvEZ9bRvnMmdt\npo69jjkj3CW9ETgYWF/Si4qX5wO/q2i+F3AQ8EDA8MWxDPhpl3ajOHZmnz2+Azyk6HcCuANYIOkd\nZnZSqd0Z+ELrCcpJ4EXMZD3g5cBGuJChOM/Da459FJ7g7T3AT4DjgW36xnR9cazVgTWBa4rzvdHM\nHtbfoaSvFmN6Cj72F+FCvopT8RvLNaUxzRLEwGfwWIujgWOA7xXHKPP64vcjgNWAXwCPB24Hnl46\nv1cDewObSXpu8fIqwKpApXCXdBDwDuBOfO4nzWzjiqanAN8vjaeJI4HrgJ2Kcz2BIgiwOGbWdSd9\nHYEn/ntLw3neUPzeFbgKD1B8EtB/AynTNpep31+AfYrf7wO+WRx/a2DnvnY5631Rw7lX0XaNyjTO\nJYlrsyB17NVMTk7OqZ+JiYmDMtruNeR2ozh2Tp+nT0xMbFj8vd7ExMSpExMT609MTPy8r92PM/p8\nQmK7c/t+/6ih7UkTExMPLf7eeGJi4uSadj8pfv+w+P39hj4vSDzPH/Sd5w8b2n53YmJiQfH3KhMT\nE2f2vb/6xMTEwyYmJo6amJjYpPh56MTExOoNff52YmJizYTzPCdjjnrXpzem8we87knrqHj/jMRz\nPDt1fBlzmfQdKs97/3qtaNe63icmJq6emJj4SNVPTfuka5Q5l41rs8vY+3/mzM69xKWSDjGz90k6\nE/ikmZ1d0/ac4hGs7dE3td0ojp3T50a9RzgzWyxpIzP7u6Rlfe1+IunZePqH3rH/UtPnBpLO6DvP\nqke6u4o+VylSRtxV0x/Aw83smqKv6yrUAD1WK55afi9pQ2BhQ5+XSdrYzK5raAPwd0mvA9aS9HLg\n/xraPqj09wJ8NzmFmd0N/FnSw8zs6pbj9rgK+EdCu0uL8/sNxaO1mV1e03ZBcX2QtBDfcVaRet1T\n1xHAjZKO6DvPoyrarS/pEWb2p0KPvW7NsSF9LlO/Q8DUk9ZF+JPgP2uapaz3O/EnhlRSrxGkz2Xj\n2uwncewzTySl0XLm/cCOxd+74Y/ddcIw9dE3td0ojp3T568kfQXPz7Mt8L+SdmP60bjHRsCnmRZs\nk/ikV5H6SPla4L+BDYG3Mf34WMUfJJ2IL7ZtgV/VtPsEPuYDgDcDH2zoc3vgL5JuKv6vU3e8Gn+k\nvxnYCn/Er+MY4HeSLsVtFB+rabdY0i74F34ZNAri1YBLJPXyJU2a2e4V7R7HTF3vJFCnJ303/sj9\nIOBCYL+adr8vXfenUH/dU9cR+M0KXEXSO88q9gNOK+o1XAu8rqYdpM9ljurqP/Dr9DLg90DVNYe0\n9f43Mzs+4Zg9Uq8RpM9l6tqE9LHPYC4K93vMbAmAmS2RdG9D29vM7OCEPlPbjeLYyX2a2RslvQDP\nk3+imZ1R7JK+3df0UWa2WcKxAf5iZt9PaPdWM3t5Yp+vwXXnE8DJZnZ6VSMz+0axeB+L6/T/Wteh\nmU0kHns/4D0946Ckj+EFYao4HRcgjwD+CGxQ0+4BuEDo0SSIP55ykma2Y/n/IrFeHQ81M0lahBs+\n64THa5m+7l9tuO6p6wgzO0TSg3A7wzygSggDrGdmjy2N52XAL2uOnzqXqd8h8Gpu+5aO/1Gq7SIp\n673uplhJxjWC9LlMXZuQPvYZzEXhfpGkL+O7jq3xR6E6Uh99U9uN4tjJfRaPcWsA1wMbSnqlmZ1Q\n0fTiQnVSPnbdo1rqI+Xmku5vZrVqjj4vh96NYBVJ59Z4Cr0JF0brA8cBmwJv6m9XtN0GN5ROfYHM\n7NkVTd8MPEXSS8zsDvya9vc15TmBGz/Bd/kfo8JTKFMQXwI8m5lf9B9XnMPrcA+LXrt7cKFcxWuB\nL5nZTTXv91gLv/5/BdatWx8Z6whJx+C7+7WA+wFXUjKkS9oZ2A54haTe0+F8YBfgazV9ps5l63eo\ng9G7db2b2dt6a71QG65dtP1qjUdR4zXqo3Euc9ZmF4N/mTkn3M1sX0m74l+Er5nZrN1GidRH36R2\nozh2Zp+n45b2spdBFTsAz+s79sNr2vY/UtaxOXCLpJtx1UTVo3TZy6GX5XOSai8H8BvADsAPzOwz\nkn7RcPzP42qcl+ACtE7AXgr8D3BW8cWsukZZnkKZgvg03NaxBW6XuLOm3Rtx74eD8R3aW2raAawu\n6Te4WmiSelVP6vpIbQde5/jRuJfHQcDX+97/Lb6r/AfTeupl+A2+jtS5TPkOnQT8oDi3D5eOf2NN\nn63rvfg+HowL1ffiqtLHF5/5r4qPtF2jMm1zmbM2c8c+gzkn3CU9ANdRC3iApPPqfKOBfwc2M7Pf\nFBP23UHajeLYmX3ON7M9at6bovd4XPR9i5k1qY++2NZf0ecmCW2OBo6WtJeZHZvQ7XyKBV78f3dD\n25vN7CuSnmVm75c0azdcOo8zJN2OfylnBeKZ2U+Bn0p6gpn9Gjx1tZnVGbdyBPE8M9tH0rH4rqru\nxnadmV0vaaGZ/UjS+xr6fGfDe2WS1kdGO/D1MylpLTO7WX0xP4UB9/hC1w9+vbfFdb91JM1l/xNT\nTZue0futuGC8B98dnwBUGcFT1vu++NMXwGIzO1DSusA5VAv3xmvUR+Nc5qzNDmOfwVyMUD0Z3xm9\nE3/8ObGh7UlM3/kncN/sQdqN4tg5fV4s6cmSVpe0Wp16QNLTJV0JnAX8SR5kUcfJ+C7ra7hP7Zdr\n+txC0i8kXS/pN5Ie39DnLyVtW5zrDyQ9s6bdl3H/5kfKPRi+2dDnMkmPBtYs9MPr17T7MUwFfOyN\n77br2EzSyyW9Crhe0ttq2l1nZtcDC83sRzR7giyVtAb+iD5J/QZpSXHTnyyeDDZs6PM3uO/yO3B/\n8rriNknrI6MduPH1bcB18riE+9W0+yRua/kAbtw7uqHPpLmUdJWkK0s/TWrQrwNPxIXvPbgNp4qU\n9T7fzG4p/u6tpyXUP4WlXiNIn8vUtQnpY5/BXBTumNkRZvZbMzsc14fV8WAz+2LxmU8w072oS7tR\nHDunz6fhC/My/LGursDJh4DtzezxuD70QzXtMLNtzewpZrYN/vRQ5572Wdxw8yBcX/q5hvM8At+F\nH4x/0et2pefgO40DgHeZ2X839Lk//uj7WfwLOePJQNNh5V+WNCGpF5zTH7xVZr/iHPbAg26eX9Mu\nRxAfBrwV93i6hmk1QD97A3/GdaMT+G6xjmOBv+DX8s+4faKKpwFfAW4CLqd+faSuI8zsIHw+DwS+\nBLygpumTzOxIYFszew4eJFVH41yWeBRu9N0c+E8qbBcl1gS+BTzEzD6G656rxpOy3u9Xav/+0ut1\nfaZeI0ify9S1CYlj72fOqWVwH9n/AH6I361uKb7IVQbLSUkTZna5pEdQP+jUdqM4dnKfZrZlTR/9\n3GuFD7GZ/VVSk096mSXU6+bnmdlviz7/V1JTMfO78Ejb1czsQtV7AB1jZttT8sevw8x+J+mfuNF1\nV9zdrsz+xc+R+I55feDeYkx1ni2963Kbmd0tqW697w08Ev/yHkCN0bc4z1MBJK0PnGJmt9Y0vRPX\n6f4L7qVyaV2fwAZm9j/F3/8r6SU17d6M31yuw9VHf645x9R1VJV24mFUu0yuIumJuJpgNRpiFhLm\nsteurKY7X+4FUsdquED8laTN8SenNurW+88kvcnMpjYwkvbBnR5mkXGNIH0uU9cmdBv7nBTujyp+\n9i691vtC93+J3wKcLPe9vY5639vUdqM4dmufKvLPSPoZfcYvM6vyX79V0r64ymMH4O81x6bU5zw8\n7Pqcmqb3yj0jflr02aQfn8T1fmfIXeJmeRgU3CEvml72H698pFS7Z81JxWP7k/HH3iNwP/9DGs7z\nT7iv8VsLnffFNe0+atP5fw6QdAI1u0h5bpDD8Zv5KZKuNrNjKpq2haGXuZ+kB5rZ34r1VLdR+CA+\nN1/HDWzn4/7SvXPLXUeQkHai4AR83HvhxtIja/pL9pIqhHnvPDemPuAH/Ka7Kz7uPajxH09c7+8G\njpW0F64q/dfid2XOJ9KvEaTPZerahMSx9zPnhHudkUXS+yvaXoRbufvbvs/MDunQbhTHTumzF9yT\n6me+B64S+TBu2GoK5Cn3eZeZ1e049sKDmD5W9Pmahj53w10Qv4cbIl8OIGkTmxnpeUHxe6OGvsrn\n2eRZ81/Aq8zsn5I+BDwHuKI4h29VdWhme0pa28xul/RLM/tbcZ67mNnpqs7/M49mY+GHivM8FfgI\nfQK2xCPMbG9J25vZtyXV+eKDC40LJC0B1sFVWVUsM480xczuknRb3/t162h16rmfmZ0rT55mdU+B\nhUqx59ExZXDuX++l46d4SZXVRb8Fzqw7STO7gOn1VN5xn2Zm5XxFrevdzO4EXl4I34cB15pZbQwG\nideoIGku29ZmX9vUsc9gTurca6jNplbB04bcbhTHnuqztABXx/WUZ+I7hTo+b2ZvN7OdzewdVu99\nA666+G882dgRkh5W0+524Cgzewy+O1lS16GZ3WxmZ5jZpJn90Mx6Tw5f7Gt3CB7o8g/gfyuEQJk2\nz5pVzOxiSRsDa5nZrwuVSGMRYDO7vfj9t9LL+xWvHVbYGD6M77B3xG8WBzR0uawY76SZ3QX0C9ge\nqWHomNk5+E1yJ+AVZvaDmqZXFLvdDYqbxQyPidI62s3Mri5utAtpdlvMSTtRRdV6T/WS+hJug9oa\nd0NMSevQz/37/k9d75jZDWb2855gL55cq0i+Rhlz2bg2E+kf+wzuS8J93gjaDrvdoH2egKsbtsEF\n5XE1n11d0mMlraF2b4ijcQ+d7fDHyapdJrgA6O3w/o57A+UyY0yFINoTz4XxKklNBtU2z5qe6uc5\neMg6klal2UCddJ7AM/FH+A/jhtJPNXy2UcCWOBjf1W+FP3435Uw5AniZeT6Y/5D0mZqm+xTHOw83\nJtc9XT1G0j6FB8aXaTbmvhafo5S0E1VUreNUL6kjcZ34OfgO+guZx4bZN/fU9V7FpjWvJ1+jjLms\nI0fWNG5s5pxapoHGgXRsO+x2g/Z5h5l9r/j7u5L2r/nsBB6oUu6rzlC6hpn11BbflPvMVrGWFXnR\nzezLkvauaddE/5h2MLPtAIpFfmHdB83sc5J+ADwGuMzM+l3Ivi/pfOChwAsKI/bncNe3Qc+zl272\nYDP7qqQmldQ+uP3kPPxpp/I6mdmP8VoRi6w98vQJZrZP8bn9JFWlOsbMluI3/zb+H74rXoR7udTa\nT8zsWkmvpZRoK5NZ6zhhLntsama9J9hvSrqgpl0Oqet9CnmqgFvMrPKmnnmNkuaygRxZ08h9Sbiv\nDFwj6WA8yf8TgbslPQvASpkkzWyLjD4XSNrCzC6R1PS5f8r95S/EH5ObjFuprKrpAI15NCxcSVvj\n+tI1gB0LvfIbeu+b2cclfQtYYp4R8RG4Gum0YZwnbiT8iaQdqY+oBPdUuI5pI/YLqQjDl7tUvg5Y\nQ0XQi5kGwlbbAAAgAElEQVRtXteppA3M7BZJ96fj97LPkLoqHln5w+JaVhpUC+PxdrgarjdHT+hy\n/FKfjXNZYg1Ja5rZnZLuR6KLXwup651iro8BbgXuL+k1hVqlv13WNRrGXA6D+5JwXxnUMpN4IqGH\nF+/fgIcoT1LKJFnsqt+C+78CYGZ1O/c3454BD8KFUp2xbm9cV/lZ3KDYlPWvjv4xnYy7uF2Ie7k0\n7bKPx93Nau0HZlZOcfwn3OOgC/3nuSeuIz0Gz5nyqobPno1fn3JGzqocK/vh3jFN9pAeH8ADwxbj\nAVRVgjCFVIN8GZnZIzoeD6rXcetcFnwG+K08udzm1MdLNNF/jNT1Dm6A3r7YLDwY+AbV3jU512jQ\nucyRNY3Xd84Id9Xnpu7lKn9ll7ZF+/vhwkq4f/aR5gmCOrUbVZ+FBf3xvbYNj7Ovx3PL/K3m/TK/\nB15r02kSKitBmdkVwK7yxEZ3F8KzEUnrl4yp4E8c5T4PlXQW7gp6bMN4AP5oZse1HTMXSZviutSL\ngb+aZ+n7ZN95/hHPzAc1ybBKLDGzPRMOfTFwjTWnhugd/zuSvgcs6jOuZVEYUCtL/FGv879Iksys\nMb+56svHVbkPJs2lmX1J0k9x//Ebrb4mQW0ZSDN7cV/TpPVekBovknSNin6S5zJlbRbtUsc+gzkj\n3PFdXc8/leLvR+J3v9WtKFLQoS14VJ/hXijb4cbKPQZoN5I+JX0Q932/CHhz4epUleviZksvLvEl\nPO/Nb3Bd/cso5YMuVDHH4E8Me+Fh0zdJ+oKZVRq4JD0ND6ZZRdIpwNVmdoyZfbCv3WuACTN7u6Sz\nJZ1oZnXpF06Vh3ZPuSFaQ+GGFDTT3/p4fI28yZqTt7VxljzgpXyeVXrVc4ErJf2J6XJ8/fVB/xX/\nMr8Uz9dyijxnzh5mVmufSCAnT/oS4BfFcZvKBlb67dd8N5LmUu7fvbqZHSTpFLlLYF1K5VRf88b1\n3kdqvEjrNcqdy8y1meNnP8WcEe5mtm3v78L74wO4J8S/D9K2YAMz6yX0Ob3YLQzSblR9/juwtZkt\nK3ZKP6OUyEjSR4o/Vyt2xL9mOq3pQTV9zkiTIC+6W+a9xTHvKbw/dsKFwo+o917oBdO0+Xq/numU\nvM/DF2adcH9j0V9TZaVcyv7Wn1ZzVspUnop7FfVcAOtqvb4OFyxN4/kc/iS3VNIn8RD83+MC6ukD\nnGNOnvRnAOsXxtomcvz2U+fyBWb2RAAze2lhMK8T7qm+5m3rvUxqvEjKNcqdy5y1meNnP8WcEe49\nJG2JuwB+nxZLf0bb30nazszOL4wsV8vd6ObZzDzoqe1G1ee1uF/yEtwg1h+A0XssvBX3Cf4/XLg2\n+cS3pUm4xzya7uHF31cAqLlQSS+YZtKqg2l63Nv7QhQ3jyZPgFsadm1dyclKmcraZvZvCe2uBX5h\n9Zkoe319S9IGeJGHc8CzBA54jjkl/i7Hg8yagnggw2+f9LlcJmk188C0VWl2zU71NW9NCyLpIWZ2\nLT7uo5k2ki6iWo+dco1y5zJnbXaKRZgzwr24CAfhj1CvNbPzhtG24KnAsyXdw3QWwcuZ7UKY2m5U\nfW4MXC7pt7iB6Z8q3MPMkyEdX4z/F8DLzetZ/hS/wc3S1RW0pUmYlOe1eB6eZRJJa1My1laQ6uvd\ne1K5CPf+qYwkLbhZ0pHMfBpJyn7XwFfwXfUmas9KmUqq4FydaWNhr12/eqD3JX0mhb1C0jyas1Km\nkFPibzs8X8zNvbY1apnU8nGQPpdH4NfzEtwu03RDSC0DmZIWpD9PEUwL+KrrlHKNcucyZ23mlMCc\nYs4Id1wFsQmuhthcniAHqFwYOW0xs0ennECvXXH3/bvVl8hKbpt67IKXJra7xwqDp5ldqerCx73j\nt6VJOB5P7LUq8Ay5QfUkvCBGHWVf7zuo9/X+kKTv4Abi482sKX/GFcXvtqIiyZjZ/0j6Pu5vbXi2\nvkHZsvjpUScQKpNgaWaKhkvlVbq2Al4j9/D4AH2G6Vwso7KUmVUG7mh2GHxq+ThInEszO0bu3vpw\n4E9WFPWuODYkloFMWO+Y2f7F7x2Lp5GHAVdYTRWyxGuUNZeZazOnBOYUc0m4n1H8Xov2qMOcthR6\nt/5ESlVl4VKTQiW3TT12waq4gC+XJ6vaeVxd6N97pfvaHqmreFpxLidI+ibuIXN3sSj3NLOm3NpH\nAvub2REAko7DPSlmIOmxeHrVq4FPSfqI1YRjW01qArXkz2hCbsx8kZmdUvx/LvU72CT6BWfpWP05\nherS136xdA5vwyNuP2lmvyzUdr/HXQQ7o7zKUnXsx8xAudRSgFlzWfTX32f/sSGhDGQLs9IkSNoT\nr7PwB+BRxRy2eUvVnWfWXGauzU5jn0vC/Slm9uz+L8kQ2oLvNMEX+hOpqKNZkJoUKqdt6rHBw7ZP\nw6vHX0f9jWvPot/n4guzNp97A/NgysXtLuCrknYDbgEWq6YuasGz8Ajal5i7e9VVcToCzwZ4CP5Y\n/wm8bFgOjfkzWrgTOEzSe8zsh+T5EOeSmlNo6hyK3e/35IFrvzR3Fb1EiQWQG8ipLNV6ngWrK60U\nYBOpc1k1TyllIHP7fD3wuMJutBa+y84R7oPMZc7a7DT2uSTcN5S71T1VmlnHqmIR5bTFbIZ/6mXy\nwrNVpBoKk9tmHBvgdjP7qKRNzWwv1XjWmCes+nRDPyn0nibKdVENX2S9cPw6rsB3Kt+WtAeerKmK\n1LzvKefZhZuAF+Nh6BsCbR4hg5B645gajwYsgNxATom/1vMsSC0FmNNncjtLKAPZ4di3MJ2zqOeg\n0KnPDnOZvDa7jn0uCfdnAo/F/T1rc0V3aIs8L0SPB1G/I041FCa3zTg2uHHzgcDCYifRJSlWFpZf\nF7X3uV9JeiWecKyu7Fhq3vdRMc/MFkt6Hp4D/bEjPFaXm9BABZAbyKkslcpvcD/rzXGHgA82Nx8u\nhZrjWLwC1N+AvVpUh019fYVp75hfySOon0C3rJQ9cucyeW12HfucyQppZv9nHgyyNX7TEe6W9PNB\n2hY8qPRzF+5/XMUbSMu6l9M29djg6osX4r7gV5KvwphCRbWnBvp3mudIeoek9/Z+Gj57MkylA9iV\naQNaP7vhBtvP4juVqbzvbec/JA6EqdSqL6C6+PHypvwof7eZ/RlXsW2Eq7f+FU/VMAh742szpcRf\n63kWpJaPGwZVT0E5ZSDb+jwC3xQegOvNvwK8HY/56NRnh7nMWZudxj6Xdu49DsHvUJvhvp8H4qHU\n2W01nabgi7M/Wsl3zOxZw2jb4di9SMdeQEyT22AKJwJPbjBI9oeNt0Y1StrKzH6JR16Wx/7ZqvaF\n90PP+F0OJikbFdtIyc3Sf557m0fX7iKpqd7lMKlUy6glRUPB1/HQ8t61rwuMaqRvTsB3pmfRnAgt\nNQw+tXxcE7PmstiVTuDpNnrFO6rcepPKQKqoRlX6/wQzeyWl9d4zdkt6PrCVmb1P0pk0pHpOvEbQ\nMpcd12ZOCcwp5qJw397MdpD0QzM7XlKTT2db216iqg3w4KBL8MK9N1Cd1W1xccEvZ7osXF3wR1vb\n3GMj6cPAqykFiGQajcpcKelGYF1J11FK1WBmG9vssPGUqMZn4sU3+m+2MxKbJTBLEKpj/owaemOr\nLQzdFSXmWFFiioaCB1p9GbwcevNSTs3R+79yfpQeBp9aPi55Lgvj47/jcRAHSPqamX264tjQUgZS\ns6tqzSt+fgdQsd7BN4c976fd8EItZ1WMJydVQNtcdlmbOSUwp5iLwn2BpDVwfeEq1BvrWttakaZA\n0mnAK83stkKX/ZWa/h6AV7bv0RT80di2w7HBA4k2sYao3FTM7BXF8Q8zszcmfKQ1OMeKqEMrEmdJ\nejTwT/PEWzlU6ac75c+owszOkrRlccNfDVeZ3Y2rFgYlNcdKaooGcEP7xlYksRqAjTO9yCA9DD61\nFCCkz+XzgO3M020swFWcdY4CjWUgzeww3PvkIDP7SMXnq7jHzJYUn1/SYPDPSRXQOJcd12ZOCcwp\n5qJw/zTwK/yR8uc0V8VJbfsQM7sNwMzukPtyz8JqfJgHbJt07ILf4Jn8hhEm3+NAeUKyBwPfwXdR\nVTry1qhG1ScZO9pq4gEy6JQ/owp5kZPdJG2H6zI3ofC1J6+MWRWpOVZyPK+2B/4id3XruRl2eWLL\n8iIrSAqDN7NzJD0dd+F7uHmwUB2pc3kDHgl9O646avKh75WBPKfYSc8oA1lyXLi5z4mhKdL5Inng\nUS9epM5ImZMqoHEuO67NxrHXMReF+5vwcN9NgauKR95B254t6ce4SmFrakJ95YWXX01pZ1n3Jcto\nm3TsgkuB6yX9jensc3V52lM5hulC1n8r/p/lk20erbcuHq33p8LQ009TkrEc4V6lnx60lmeZlwJP\nwedmd7ziz/9pOJV+UnOsJHtemVlugFEdWV5kBUlh8PLycVeY2X9Leo+k/zCzOmHUOJeaLiryAOCP\nmk63cUvDeX6V6YCgXhnIcs3Tuk1TU5T5vnKvIgGn2HQFp36SUwUkzGWXtdk29krmonCfxA1uhicW\nasp4mNTWzN4t6Ym44aYpDH5n4GGJapGkthnHBtf7/SvDzYy4gZkdK2kPM7tANYmMJL0Y11kuAL5W\n7Dj7g6O6JBlLNSp2yp9Rw21mdq+kJwBX2nRk3zCCmFJzrCSlaIAp9dYRwHr4F/dSK0oe5lCM8yeS\ntjavGdB/nKoI0dQw+JzycW1z2aWoSGMZyLIaqng6noryrutQ0vq4G+/1wHqSDjSzWWkjMq5Rylx2\nWZudSmDOReGeoxdNaiup3G4neRKva4DDzKxswc9RiyS1zTg2+O7ujmHo3PvO4VHF74dQHyyxP64X\nPROPeP0lsyNfs5KMZRoVO+XPqGFS7gq6J4XXkdzbYRhBTKk5VpJSNBR8tjjXo5l+0soW7j2qBHvB\nrAhRZYTBK718XONc2nRRkXI5vh51lYuSykBKOgbPp74WLrivpN52cxoe4b0F/nRxZ02fOakC2uay\ny9rsVAJzzgl3KzIfDrnt/fCSbD/FJ/pJeHDB8biPaY8ctUhq29Rjgxd//pOkK4v/J1ss7ym8GX+6\n2Qx306r78txrnltm0swmJd1R0SY3yViOUXHQ3CFlDsZdQf+G2xyeVpxnamK2JlJzrKSmaAC8ElZx\n7W9q0c8PQtWNKDUMPqd8XOpcppbjg/QykFviXmlH4gFFX2/oc56Z7VNswPamPio7K41Fy1x2WZud\nSmDOOeE+IhZZ4T2CV9I528zeU/FomaMWSW2beuxen8PmOVYqbtLAefLIvYcU+tVZHgGWn2Qsx6g4\naO6Q8nn+Avfxf4iZ3S6PQHw4s1Msd2F1peVYSU3RAPB3eSTpWnKPpWGq5dpICoO3vFKAqXOZXFrR\n0stA3lJsTtYys5s1067cz1K5t91a+FzWycOcNBaNc9llbWaMfQYri3BfR9KjzOyyQkWxUJ6qtz+8\nP0ctktq29diaDmzYh9m7qzp7QyrPlfQpa6nlaV7q7Dl4Du4/VOl8lZ9kLMeoOLSo1eJLsDHwCUnv\nKF6ej7uSNSVuSyE5x4qlpWgAN8wfBNyMp4xtyj80bBrD4NWhFGDGXLaW41N+GchfSXobcF3Rd9N1\nPwy/AV+CF1ep27nnpLFonMuctdlh7DNYWYT7m4AvFTvNa/CsebsxnQOiR45aJLVtyrHrAhsGSZrV\nY0N8oV/F9E5z1nkWX+IJ/JFzc0mbm9kn+ppVJRm7FzcaVlE2Kt5Os1FxaLlDcGPWK/Aw8JcznQzt\n8I79lUnNsTKVokHukXFYQ5934BlBe7rnR+KBPcOmSv0xFQYvD8p7c9/72aUAM+YypRxfVhnIYpOy\nEM8T0wuQqmMSX5OLcdtZXTR52zUq0zaXOWuzSwnMKVYK4W5mF0l6BjPd/H5Z0TRHLZLUNuXYZtaL\ninuS9YVO40Eyg/D8xHanA9+gQf9p00nG3mBmKYJyLTzgp+cp80LqU6r28mf8VtLjcGG4XeK595/n\nT4GfSnpv/05wCBwL/BgXbk/Dc6xM2U6UmaKh4Azcz3sx0xWBXtT1BCU9FBcgU4ZKM/uAlSJElR4G\n36UUYOpcppTjy/bQsiKuBFeJ7Uy9cfo9uPC8UR51+21KEaoZ16hM41xmrs1O3mk9VgrhrjQ3P3Bd\n2oywaepVCUltU46t2aHT4Avj9wyITVf96R1r56rzBK4xs/cndrs7abvgs/Ex9HZmk9QL9075M1p4\nJm4IHCZtOVa6pGhYw8xS88Gn0JoniPQw+C6lAFPnMqUcX5cykGUqqygV3GJmNxbHvUHSrX3vd0kV\nkDqXKWtzoLGvFMKdNDc/yAuBT23bemzrFjrdlbrF/m1JvfDm3nnVPTXcIelTFPEFRduqKMAlVqQq\nSKBT/owWysbP3nnmFpjopzHHinVL0fATedDPH0r9DFISsDVPkKWHwXcpBZg6l/3l+KrUkHUeWk1P\nQshdVW8xs6YI99sknYU/iT0RWFNe4QwzOyjjGpVJncuUtdlp7D1WFuGe4uYHeSHwqW1bj61uodNZ\nJCz2l+MLabPi/yZ9fy+abqOWw54laR9m3jDqgl865c9oYRgFJvppzLHSYARrStGwEZ5Ko/yEM4gL\nbGueIKWHwdeVj2sSMElzaWaHqCXgyDI9tCTtiF//W4H7S3pNT5VUQTnSdFapyoxrVCZ1LlvXZu7Y\n+1lZhPt5xe6j1s2vICcEPrVtyrGb8s0MRMZiv9vMkqJCiy/lv+HuWxfihsUqngqsznS6g6ZUtp3y\nZ7Twa/xLtDFFXp1BO7T2HCtdUjQ8ysw2q3mvC615gkgMg7eifByApAfg83Iqnquo7ukiaS6VGHBk\nZrdKWihpFwo7gqQtap4sP4hni71Onp3yG0ClcLf2OJkuqQJS5zJpbWaOfQYri3A/HC8q8Qc8Mqwu\njWxOCHxq29ZjWxE6rekc8MMkdbFfLelAZuo/69LEfoS0nPtrm9m/JZ5np/wZLRyLC6an0ZBXJwe1\n51jpYgS7uNgglHfa/+x6juZ5gjbAnx6utOqcS1lh8JIOx2v29tJHNz1dpM5lTsDR6cWxy3nSq7jX\nioyMZvbXlqfvNrqkCkidy5y1mTr2Gawswv1LwPtx16uDcL/dWVkdzexaEvNeZLRNOnbByfjEzccD\npP6IZ5kbhNTFviruCtlLfNRkAEzNud+qHijRKX9GC0l5dTJpy7HSxQi2Q9F+qg8GCLiS9FLcrvMH\n4DGS3m9mJ1WcZ04Y/Nb4k0pK6HvqXOYEHM03sz0Sjn2rpH3xJ8QdmPbU6kKXVAGpc5mzNlPHPoOV\nRbgvwyf73Wb2VUkzdICSrscnYXX8S3gN/th5k5k9rGvblGOXsVIkqTx/xzD07UmLPcPwCek597cs\nfno05cfvlD+jDaXl1cntsynHSm6KBsxs2LVd9weeaO6XvRA3fvYL99ww+CtwtUBl/pU+UucyJ+Do\nYklPBv6X5h3xHvjYPozr+/dKON86slMF5MxlxtpMHfsMVhbhvirwCdySvSN9ZcfMaxMi6STgQDO7\nRtLGVOSHz2mbcuwGljCccPmkxV6oZN6Jf3l7uXLqwv8/RUIefevLeV94HNTRKX9GC6l5dXJozLHS\nxQhWCMK3MtMvPbUMYRXLrEjZbF4kZtbTmuWHwf8Lrrrrebg0BfglzaXlBRw9jZkxGzN2xMU4rsUN\nmkczrTpaRIdSjcX5ZacKyJjLnLXZOPY6Vhbhvidu2DoG2AV4VU27h1tRUafQUTfpwFPbph67nOd6\nHr4ov99w/EY6LPaX45V8WndmZnaKPAXqI3FdZGUebnmOjf2Z9oa4h2m1T3+fnfJntJznpbjBDkkP\ntepSa7l9NuZYUX6KBvCb41to9kvP4UpJhzL9tDbrWio/RUNdHeNZ5MylJQYcmdmWVa+X2L/4OZJp\nnXRvzXe6UXa4RpA4lzlrM2HslawUwt3cx7jnZ1wXRAPwe0kn4juIp+C704HaZhwbZurw7zKzG1ra\nN5G72K/Cd1CtFJ4yC3Af76/Ks+V9uaLpG/EQ9YPxwJq3VPQ1UP6MlvN8O+6Sdn9gT0lnmtn+HftK\nzbGSm6IB4C9m1vlGXsGe+G55J3znXFUtKikMXp4bqM6Ad1Bf20HnclYMRmnDM4vyk0NvXgtj8oZ4\nRPgVNliW0S5pLJLmMmVtpo69jpVCuGfwWjxEfgL4qpmdPqS2qfQWUc/lCTPrpErosNhXAy6RdEnx\nf122Q3AVz+5Mh5V/Dc+n0c91Zna9pIVm9iNJ76toM1D+jBZejO9czzSzzeV5uLuSlGPF8lM0ANxY\neOGUDc+D2FtWxz0sTsV9zDemLyrZ0sPgc6IzO82lmmMwsnL8S9oTVy/+AXiUvJ5s26aqkoxrVCZ1\nLlPW5kD1DYbhPTBOrIXvRv8KrCvP6DeMtqkcj/u7nlX6GYhisZ+HVxD6maSX1TT9OL7TPqL4aSrT\ndide/3JpoZqo29ktkSfNmixUNBtWtJnlOmieaTMpf0YL9+I76N4TUGrIehWzcqyY2V+p/w7lRMJe\nhVcDeiAe8/DA5uatfB14Am7ruYdmw/wzmzoys+PN/cG/hMczXAX8Gagy6GXNpaQd5Yn3zsFVSTtV\nHP9q8xQaq+LX9FV40ZO6bKmvBx5nXnFqK+CApvEl0niN+kidy9a12WHsM4id+0xy/Ek7+Z62kJzf\nOoPeYr9L0lq450TVTqbfx/YeeQKqk212ZZ9b8XQKR8nz4txYc+y9cb38gfiXbN+KNoPmDmniR8XP\nHvJ0Cd8doK/cHCupKRoAftj3/z0lm0kX1sSTYL3FzF5ZqNHqWF1pKRpOw4XMg/FNzXV4bdEyuXOZ\nHHCEPxmehrsGX8fsdN09bsFvaOBqxmHkxk+9RpA+lz8ifW2mjn0GIdxnkuNP2sn3tIXW/NYdSF3s\nWxbv9ypGPRTfgTwbV0GUeRnwCDP7fWF0Orqmz2V4oYN7ce+fqyraDJQ/owkzezf+xIKkX1TcpHLI\nzbGSmqIB3Cf9gbjd5vH4rngNecqC/+pwrqvh4fG/krQ5/pRZR2qKhg3NbFtJX8Bv0lVCOHcucwKO\nbjezj0ra1Mz2kjQj97q80EzPYeBXhWfLE0i0I7WQk8YiaS4z12bj2OsI4T6THH/STr6nLaTkt06i\nw2K/v02nhD1SXjHqPyVVGQIfAOysmRkRq25CX8dVPC/Gb1hH4TeLKbq4DqbS75ZW2DC6uhhm5Vix\n9BQN4GquxxZPV6vja+BFuLdLF+H+Ntwz68O4K2xdHhRIT9HQ86Jay8z+IWnWk2qHucwJOJqU9EC8\n2M1azN69HlHxmf4ni67kpLFImsvMtdk29kpCuM8kx5+0k+9pCyn5rVPJXez3l7SheaTgBrgdYVWq\nH6lTUspSfPZbwH5N6gEbIH9GC0NzMbTMHCtKT9EA7lZ5V3Gcu4t5+Kc6RtSa2fmFLnsdXD3TVK4w\nNQz+G5LeA/y22CjcXnPsnLnMCTg6BE/jcSKeg+bEvuP+uDjW84GtzOx9ks6kPv4kh5xUAalzmbM2\nG8deRwj3Ejn+pF19T1tIyW+dRIfF/j7g5/Kc1mvjj94HUJ3sqjWlbEGOemAUNoxhuxjm5FhJTdEA\nXpvzPNyt9knAt4r2l3Y8x3JCrjVxP/e61NVJYfBmdpikeebpAr5Lka5X0i4VnmKNc6kOAUfm2UR7\nqR6+VerrfVbkZio4hOn0HrvhQnlQx4ScVAGpc5m8NjPGPoMQ7iXk1VbeyHTQzQZWE06c0zaD/vzW\nwyBpsZsH55yBf8FuLHaqZ9b0mZoz5gB8x5GiHhiFDWPYLoaQnmMlNUUDZvZBSafju/xjzexSuXtg\n1dNXCjkJuZLD4Is1gZldUnp5P1yYl2mby2EGHM1yBDCzJcV5LlFi1aI2Mq5R6lwOY202JsEL4T6T\nD+HBH/vgVu9Zrlkd2yZRdxeWdFrh2tWFxsUu6XNm9ib1BUwUOsC6QImUlLKY2QVMGxY/V+q7ajyj\nsGH0DLhNxSBySc2x0pqiQUUZN80MFNqyuPaDFEbPScg1aIqGqgyJjXNpww046j/+RYXh+2f4jXhg\n2w0J16jDXA5jbTZlpwzh3sf1ZvYzSfuY2XGS/t+Q2g7K/Qf4bNti7xV4Tg6YsL6cMR2oGs/QbBil\nx/5hGdTKJOVYsbQUDV3KuKWQnJDLBk/RUCWUkuZSwwk4mnF8M9tXHlsh4BQz+1b1x9JJvEZJcznk\ntdl4QwjhPpO7Je0ArCovxLHBkNoOSucdZ8Ji369hZ9cfXv51M3uJpjNjlo/TZLTrp8rTYpg2jKHn\nGSmRlGNFaSkaJuVFtK8f8JxmYJ6Qa23cN78xIZeGmKKhdPzUuUyNwUhG0vr4zex6YD1JB5rZRwfs\nM+Uapc7lKNfmDEK4z+Qi/Jp8CN/RNl2fnLYrjITFbql9mdlLit+VlaNqjGtt5zdQ/oyaz0099tcc\ns9EQVfOZ5BwrBSkpGupuFE259FuRBwSVi7c/jOlIyH4GTdEwpRroMJetMRiSJmrsObOOX3Aa/iSw\nBX5zS0lR3EbKNUqayyGvzVDLtCHp1Xg05WZMBxCtQoVPeE7bOULjYrei1Jg8svBJNNSzTKDKuNbG\nQPkzOtKlGlOu6mRGigZV+4VX5tCX+4cPQk6h9+QUDXLf/gngd2bWux6fLDVJmkvlxWCciKfdrbM7\n9af9mGdm+0g6Fv+eJgX8tJCSKmBYczlrbfbsYqX/TzCzVzJ77DMI4e6cBPwA34F9uHhtGdVh9Tlt\nh0WnfNQFqYs9Jby89ViJ7abGY547A0mPxDMulm8uw8jpXkXqeU7R4SaYmqIBSR/AVRSr4YLjctzb\npSs5hd5/REIYvKSDmVbxHCDpa2b2aTP7dq9NxlzmxGBcKelGPO6i534KRb2BCv330sJLaS38BjIM\nGfcjElMFDGEuy09Cb8TjANaX9KLivXnA7wDa7CMh3PFgAzwZ0mtbmma1zUWey+UVzEz0/wGbjhzt\nQqxMUyAAACAASURBVOpiTwkvb6PfnzlnPJ3yZ3RkEK+Z1JtgaooGgBfgAU+fwnfCqdkk60gu9G7p\nYfDPA7Yzs2XFDe484NM1bRvn0jJiMMzsFUXbw8zsjXXjKHEYHqF7CXAtQ9i5Z1wjGHwup9ammR0G\nHCbpIDP7SGY/IdznGKmRnzmkLvbW8PIO5IynU/6MFUDqTTA1RQO459Xd8tTIV6i5YlUKyYXelR4G\nfwO+E70d35Xe1HD81LnMCTg6UNIH8Zvqd4CLrSg+3sck/oS6GI8M/mLDeSaRcY1giHMpqbeBvLn0\nN5DmEx/CfW6RGvmZQ+piTwovb6Ff3ZEznk75M6roYITLIfUmmHNju1bSXngmyY8ymOsrwFvNLNWW\n0RgGXzKSPgD4o6TfApvjxtA6UucyJ+DoGFz4P53mFADvwXPK3yhpIzz9wqARqjmpAgady/LarNPX\nJ228QrjPLVIjP3NIWuyWF16eYlzLHU+n/Bk15Brhcki9Cebc2F6HZ+E8Bc/XvTuApE16euxMNpd0\n/8SgoLYw+C4G79S5zAk4Sk0BcIuZ3QhgZjfI02kMSk4ai6S5TDGSlr1mCsNslrPDvMnJYTx9B8NA\nUn8u6MmGx7/UPr9vZv9W93/C52fV/uwzrj0Z+JqZzdK/DmM8Hd0Wv4LnXV8X32HOMMLl9FXTf+8m\nuAUeWfmP/ptgYXj7OQPcqKuufeLnrsbVFzcXx64dt6TjcJ18Yxi8pK0pVQkr2mVFs1bNpaZjMP5Q\nEYNRbncuHhl6OC4AT6xyJ5R0Gq4++jHwRHz3+6PifDtF/aZeo5Y+zjWzZ5SNpEyvzXn4JqmyKIhm\n5gq6Hx4UV+f9NEXs3OcQ5uHYG+A1KK80s5uH0O1tks5ierGvKc9YmLrYq1QYSca1IY0n222xgxEu\nt/+UHCtJKRpa6KQ+MrNNql6veQpLDYM/HvedH8Rza8ZcKi/gKDVNwjdLf/91gHMtM7RUAR2NpFm5\ngnqEcJ9DSHopHhT1B+Axkt5vZicN2O2gi71qIScZ14Y0nkH046lGuGEw4zzrglQyGfZj9dQNSPlh\n8MOoEjZIwNFzzGzbtgP0XFaHQYdr1MRk0WcXI2lOrqApQrjPLfYHnmhmt0taiIdjDyTch7zYc41r\nwxjPIAIu1Qg3DHpf3mGlaBgFZeGaGwY/jCph/XOZE3D0XEmfMrOhZHlMZBSpAroYSZNzBZUJ4T63\nWGZmtwOY2W1qDj5ZXpQFQq5xbUWPJycP91Cw4aZoGOSppYqyD3VuGPzQqoSVyAk42hAXblcxbUfo\nlJ4ilQ7XqImeWibbSGqeK2ghHsHbmCuoTAj3ucWVkg5luuzYn5bnwSVtCmyKlxH7a6FbnvKAsekI\nxFnGNap1oMMYz0ACTol5uIdA6nnmpGjIzfMyTPqfcFqrhHVwQc0JOHp+w3sripynwBlzWWUkpT5F\nBGZ2W/HntyXtjKsZGwnhPrfYE3el2gl//H3X8jqwpDcBL8St+MfjqWrfZKXw8hKpxrXk8aS4hnVg\n0FzllSS6gdYx6yZQBMnsD6zee83MnmFmH+xvOyA5N8r+tilVwnJdUJMDjvpdQgsB18VNdJgMMped\njKQFm6Y0CuE+t1gd39WdCrwGf1RbXgv45fju+gdm9mlJv2hom2pcax2PBsyf0UKSES4HJeRYaaFK\ntzq0Wq89BrwBwezz7K8SVjWO3DwwgwQcJQm4ETPIXGYbSeUVnW4xs6S6sCHc5xZfBz4PvATf6R4F\nPHs5HXs+hS6z+P/uhrapxrXW8XR0DUtlFEa4nBwrqQy11usQbkCzMLND2nTEHVxQswOOcgXcCiB1\nLpONpJJ2xJ0BbsUL2b/GzFpzP4Vwn1usie9e3mJmr5QXfFhefBnXjW8ir6X6zYa2qca11vF0dA1L\nZRRGuJwcK1VUqUaGXeu19QaUqx/P1BGnuqAmx2B0FXAjpvNcZhpJP4gXXL9Onqv/GyQk9gvhPrdY\nDTe4/UrS5vgXaXlxDp7K+DGAmdnFDW1bjWsFKeMZKH9GC0MzwnVwA81RjfQHyQxKyg0oVz+eoyNO\ndUHNicHoJOCGQaY9KHkuM4yk95rZdcVn/prqdRbCfW7xNmAXPE/8HrhgXF4cY2bb40ElbaQY1yBh\nPF1cw1IZshEuyw00RzVSqDyehwtPy3CV7D9mzg0oVz+eoyNOckHNjMHoJOAGoYs9aIC5bLIh3Cpp\nX6a9zv6e0mEI9zmEmZ0v6UpgHVydsTyDXu6Q50MxvPhIk2ogxbiWNZ5c17COdDbCdXADTdbNy7MH\nblq0eZWkp5rZ2zqcZk6R81z9eFYgzQhcUDsJuEHoYg/KnctEG8Ie+E3mw7jtaq+UcwnhPofoE3Br\n4n7hwxZwdVxQ/N6orWGKcQ2yxzOIa1gjQzbCpbqB5ujmdzCz7Ypz/QxwYZcT63ADgkT9eKaOeBQu\nqJ0E3CB0tAclzWWKDUHT6Q82wou99KJjF5GQ4yeE+9xiZAKujZzHyYxdds54OuXPaGJERrhGN9Au\nunlgVUnzzWwZ01/gQchJ8pWcoiFDRzw0F9RBBdyAdLEHpc5lig1hoPQHIdznFkMXcKlkPk6mCu2c\n8XTKn9HCKIxwbW6gXfKfnwycL88P/2Tgq4OdYlaSr64pGppUXMN0QR1FfpckOtqDUuey1YZgpfQH\nkjYEHoanmE5KARHCfW4xCgGXSo5qIFVoJ48n87E/lVEY4RrdQLuoRszs0MIl8FHAF8zsdwOeY1aS\nrxz9eKKKa2guqIMKuGGQYw/KmMtkG4KkPYF34s4Oj5LntPla23mHcJ9DFAJubTz96bAEXCo5qoEk\noZ07nozH/lRGYYRLdQNNVo1IeiwuOK4FPivpI2b2gwHOMSfJV5J+PFPFNfQ8MF0F3JBIVi9mzGWO\nDeH1wOPM7C552cJzgRDu9yUK1cHHcX3tKfgu5YbldPj+x8mT6xqm7rIHHM8wwstHYYRLdQPNUY0c\nAbwJL0/3buATeMxBV1JvQJCuH09WcQ3ZBbVHJwE3JHLUi41z2dGGcAtwT/H3P0jMzBnCfW5xFHAo\nnnPjJ/jub7l4y/Q9Th5rM6sMVbVP2WVnj2cYni0jNsIluYGSpxq5C/edXs3MLlRzoegUUm9AkK4f\nH0TFNYwbdScBNyRy1KVtc5lsQ5CXi+yt218VG68n4ONvJYT73OJ+ZnaupIPNzIakI05C0muACTN7\nu6SzJZ1oZqlFquu+vMnjGbJny8iMcKluoOSpRiaBE4AzJL2MaSHWldQbEKTrx7NVXEO6UQ8k4IZB\npj2ocS4zbQhHVLyWXBUqhPvc4i5JzwZWkbQNvgtYXrwer0APHoDzE+qr1gNJX96c8QzNs2WURrgM\n41qOamQ3/Nr3XBK7eNxMkXEDgnT9eLKKa8g36oEE3LDIsAclzWWKDcHMfly0fT6wlZm9T9KZeObJ\nVkZemSbI4rV4DvQN8dD91y/HY99rZksBzOweGnZ7knaUR56eg4ex71TTNGc8Mx77GcKNrfgCnYfr\nPn9W7KQGpWdcOwv3X687z5slHSnpdZJeq74gmD6WAesC/wn8C54DvzPFDegHePGLX9AgDMzs6vIP\nXs+03NdDij97Kq4DgC/gO+k6ejfqxwHb4XV0O2FmPy6E3DrAM4q/D2SmF9LypknNlDqXPRvCC4Gt\n8OtaxyFM5yTaDXhfyknGzn1u8VYzG2jXNgCnS/op/sj5BJqrBaXusnPGMwrPllEY4VKNazmqkV6h\n6Mfij/5NhaJTGGYhiC4qrlG4oB4C9Mrd7YbvjFNzvw+FRDVT6lzm2BDuMbMlAGa2JNUmEzv3ucXm\nku6/Ig5sZh8C9sWF+34tKoXUXXbOePbAdzofBh7KcDxbRmGES3UDPQQ3KH8ROI5mFdM8M9sHuAyv\nWrX+gOd4i3mJxLXM7OaUD0haVLjCzhBcNrOO6MuAdwAvNLMm28WtkvaVtGVxwx7GjXqGgAOWW6Hs\njCdVaJlLSV+R9GWmbQhH4DElTU8iF0n6cnFNT8TTCbcSO/e5xeb44/zNTBu3Rpo8TF6g+HXAZ4Gb\ncb3qXZLeZmZ/q/lY6i67dTyj8GwZpREuww00JxFaTqHoFIZeCCLTz3wULqgXFULxZ7hOO0nADYkc\ne1DbXGbbEMxsX0m7AgJOMbNvpZx0CPc5hJltUvW6pF2sYxrYBD6LJ7eajxcs/gXuyvV5vKZqFUlf\n3sTxjMKzZaRGuETjWo5q5DDgrcDZeHm28wY8v1EUgmhVcY3SBbWrgBsSOWqmxrnsYiSVtD5+g74e\nWE/SgWb20baTDuF+32A/mnXgg/BoM9uu2G08FXiJmd0jaZaBZ4hf3qnxjMKzZVAvg0zqjGvJgS9m\ndipMfYlPMbPWcnNtZHh3pAquFBXXyFxQuwq4IZFsD8qYyxwbQk+PvwWuAk2yyYRwv2+QU7U+l54Q\n2A64qPCUgepH+WF9eauqxo8ivHxkRrgE41qOamQH4HBgFeAUSVeb2THDOM+CzoUgclRco3RBpaOA\nGxI5bqCpc5ljJJ1nZvtIOhbYG/eCaiWE+32DYZScq+P2wk3vJcCX5VkB/wP4S3/DIX55q8YzCs+W\nTl4GTaTqqDNVIx/CBeupwEeA84tjDHquwygEka3iGtGNupOAG4SOT6qpc5ljQ+hkkwlvmWAf4BH4\nrvZ4fKf7Ehp80kfkPz4Kz5ZOXgYtJPtwm9ltZrbUvLTekxr6XGZmf8cNzncx/TTViRTvjlT/9Y5+\n5jk+3KkM2+icwv7F7yPxm9znS3/XkTSXZrYvvnlZE1ffvLmhz8PwOJFL8IRkVzW0nSJ27vcNRqaW\nKVzl3ilp58J97gfADwqBXbfbGnSXPTWeEXu2jMII19WHu0k1coU8n/4Gkt7F4Em2RlEIIkfFNYob\ndb+AG/nOveOTatJcZtoQJvGnlcXA3bh7bSsh3OcYkrYAJoDfmdllxcufbPjIoMfbGXgKsLukXk6R\n+Xhh6zqBnfzlTRjPyDxbRmSEywq2SlSNvAFXh5wH3AG8ZsBzHEUhiFYV1yhv1HQUcMMgU82UOpc5\nNoT3AFub2Y2SNsLrEbfajkK4zyEkHcy0fvYASV8zs08Xj/Wj4rfABvgX0IrXllFRQSb3y5synhF7\ntozCCJdkXEvVzRd8x8yeNYRz6zGKQhApOuJRuqB2EnBDIudJNXUuc2wIt5jZjQBmdoOkJG+qEO5z\ni+cB25nZMkkL8Lv/p0d5QDO7BjhengVyWUvz3C9vznhG4dkyNCNcB+NaTuDLYkkvAC7Hb6yY2eVd\nz5URFIJIUXGN+EbdScANiRw1U+pc5tgQbpOn4/4x8ERgTUkfKfo+qO5DIdznFjfgBpbbgdWAm5bj\nsd8p6Z347nYeFdGkHb68OeMZumcLwzXC5eqoc3TzD8ADX3p08gvv6N2RJLgyVVyjuFF3EnCD0FHN\nlDqXOTaEb5b+/mv7mTsh3OcAkn6GL4IHAH+U9Fs8dP+W5XgaLwc2NrMU1UXjl7fjeEYRXj40I1wH\nHXVO4MuOVa8XKpJDqt6rYZSFIHJUXKO4UXcScAPSJVVA6lwm2xDM7Pj2U51NCPe5wYrKBFnmKtIN\nX21f3uzxjMizZehGuAwd9TDyqzwtp3HmDShXcOWouIZ+o+4q4AY85jDVTP1zOXIbQgj3OYAVNScl\nbY0LxrIPcWXB4hGwGnCJpEuYLs22e03bxi9vl/GMyLNlFF+gRh11R9VIHZ1cYFNuQB0EV7KKa0Q3\n6hXJMNRM/XM5chtCCPe5xfF4QelB63x2IbVqUM6XN2c8o/BsGcUXqE1HPcz8Kl0jk3O8O1IFV7KK\na0Q36hXJMNRM/XM5chtCCPe5xR/N7LgVdOxf47u9jfEkUxfXNcz48uaMZxTh5UP7AqXqqDsGvgyb\nURSCyFFxrcg8MKNgFPagkdsQQrjPLU6VJ5n6fe8FM/vAcjr2sfiu7WnA33Af7Tqdb+qXN2c8owgv\nH+YXKEtHnaIakTTR4vKYpZbp6N2RKrhyVFzLPQ/MKBmSmmnGXC4PG0LklplbvBH/ct1Q+llebGBm\nx+I7uQtoXhu9ajNGc+WgnPF0yp/RhJkdX/fToa/cHCsp+VVOBJB0Wk0fr8w8zSNwddABeFrlrwBv\nB95b9wFLz3EyQ8WFB2fVsSLywIyMqifVhraf6/v/hOLP3LkcmPv0RR9DbrHm8nYjRdKjit8PAZY2\nNE398uaMZ4WFl2eSqqNOUY1cKelGYF1J1zG9u5s0s42LALNkunh3ZKjYclRcyz0PzIhpfVKV9Ebc\nO2p9SS/C53IeXviG3LkcBiHc5xY3SzoS13/3PFaOWk7HfjMuUDfDqwY1eemkfnlzxrMiw8tzaNRR\n56hGzOwVxWcOM7M3DvEcR1EIIkfFdV+5UafSqmYys8OAwyQdZGYfWe5nWEEI97nFFcXvBxa/R5nH\nfQZmdile8zOF1C9vznhWZHh5Dm066i75VQ6U9EHg/7d3/7F213cdx5/NAkLrFsaPsfJjAZf5Ugjh\nZwrMWQUlqHMKWaSYkE2WMWGmmiHRtNuCDJjORKZ/LGEs3Y/MgAG121icOIS1LpQQkehw28u1JSgt\nllF0ZUJJB9c/vufbX9Dec873nPP5fG9fj6S5p5d7z/f9abjv+/6+v58fJzJ4mG174zzfczATPwhi\nxFZWX35RD2veO1U1ZyJAU9B8YO//NsMCbR9J7hWxfZOkpcBhNLd0Uz0ce2+SbqVZaLM7AfvAh3MP\n9cM74nhmvrx8HPM9XBtz4csamur655n/YfYwpn4QxDz68ot6WMPcqS49wPfOrEDbX5J7RSStoame\nl9D0QTcDF8zo8u8ETrH90hBfO9QP74jjKbG8fGQj9KhHaY0cY/uzkq6y/ZCa07DGNuLsjmn0x3vx\ni3oE896peq+tBUoVaPtLcq/LmcDpNDMeVtP0vmflMZqZH8Mk92F/eIceT4nl5WMatkc90sKXER5m\nz2vERUTT6I/34hf1CIZuMxUu0PaR5F6X7bbnJC2x/aykWV77ceBpSf/Nnl0hf+IAXzvsD2/J8UzL\nsHO4R2mNjPIwexhTPwjiYHr0i3pYo7SZShZo+0hyr8ujkm4Atg4W/xw5w2uvAE5liCPRRvjhLTme\naRmqRz1ia+SXbA/7MHsYUz8I4hAzSpupmoImyb0itldLej3NtLn2BKNZeRL4vyF77kMpPJ5pGapH\nPWJr5FckfdL2JLbGhRkcBHGIGaXNVE1Bk+ReGdvtaen3qjnf9KszuvTJwCZJmwd/n7P99oN9wzAK\njmdahu1Rj9IaOZYmGTwxeP+u//ZTPwjiUDJKm6mmgibJvW5vm+G1VszgGrMcz7QM26MepTXyrgnH\nOPWDIOLAailoktwrJOk4mt7dJM6eHNYrwG+y714pE9m0rNB4pmXYHvUo+58/ufffBwnhyQN8+TAW\n2iKiPitW0CS5V0TSRTQLWHYAR0m6xvaBDlWetHuA+4GJ7YFReDzTMmyPusv88a4JIQ9JC6uhoEly\nr8vNwDtsb5V0IvC3wKyS4fO2PzLh9yw5nmkZtkc98vzxCSaEPCQtpKaCJsm9Li/b3gpge4uknTO8\n9uOSrqSZj91u8nWwvcaHUXI8UzFCj3qUhS+TTgh5SFpONQVNkntddkhaCawHlgPPzfDaZw3+tMY5\nFm5/JcdT2iitkYkmhDwkLaqagibJvS5X0ewJfSvN6UXvm9WFbV80/1eNrNh4KjBKa6SahBCdVVPQ\nJLlXQNJJtp8Cjgc+w54DlY9jRodlS9oCvAn4Ps286500Jyd9cNQWQQ3jqcAorZFqEkJ0Vk1Bk+Re\nh+sHfz7Nni1C24TYtTUyrPXAH9m2pLcCN9JMhfxLRm8R1DCeokZsjVSTEGI8NRY0Se4VsH394ONF\nko4FTgE22p53n5cJOsm2B3FskvQW2xsljbxDYSXjqV6NCSHGVl1Bs2hurthe8rEfSVcDf0izbP2n\ngBtt3z2ja99Nsz3pQ8DbaRLyGmCV7bH+5yw5nj6QdJvt6yU9yH4JYdx/8yivloKm06EAMXHXAWfZ\nvhw4j+YU+1l5D7CVZj+M/wR+C/ghzarVcZUcT/X2vsMBrgD+ALg8ib2/BgXNN4EPAxskXVEqliT3\numwHdg1ev8gQ2+92Jem8wcvlwHdpNrzaCCy3vcH2tg5vP/Px9FFNCSE6q6agSc+9ApLuYk+v9VFJ\nDwPn0CTEafsF4J95dYU+B/zDOG9YeDx91CaEnZKWAA8AaV/1UzUFTZJ7HW5/jc/dNaNrf1LS4cBv\nT/A9S46nj6pJCDGeGguaJPcK2F4HIOldwHm2b5T098AsNh0yrz6hvX3Kf6Bj9g7+hmXH0xs1JoQY\nW3UFTZJ7XW4C2pWiK4CvMeWtWm2fOsW3n/l4eqa6hBDjqbGgyQPVuuyy/QOAwcdJHbs2L0m/Juk+\nSQ9IelDSv03gbYuNpw9srxskhTcAFw9er2LfPfWjX24Cbhu8XkGzGLCIVO51eUTSncAGYBnNDo2z\ncgtN3/1a4EHgkgm8Z8nx9EnucBaOfQoaScUKmlTuFbG9kmaWxGLgHtu/O8PLP217wyCOzwMndn3D\nwuPpk9zhLByPSLpT0kpJX6RgQZPkXhFJR9Oclv408EZJq2Z4+ZckLQcOk3QpzeZhnRQeT59UkxCi\nm5oKmrRl6rKWZqn+GTS7Mr4ww2tfR7NFwC00+4vfMoH3LDme3rC9UtJlgGgSwldKxxTjea2CxvYf\nl4gllXtdFtm+lmZ64iXA0bO6sO0ttv/R9rdtv9v2XwFIWtvhbYuNp09yh7OgrAV+jub51Xto9mkq\nIsm9Lj+SdASwhGb+cw13Vkd1+N4ax1OjahJCdFZNQZPkXpdPATcA3wKeAp4oGw7w6gVOo6hxPDWq\nJiFEZ9UUNKmk6jIHvJ9mL++XgM+VDaezhTaeaakmIURn+xc0/1QqkFTudfkosMz22cCFNA82+2yh\njWdacoezcLQFzbspXNAkuddlu+1nAAZb7e4oHA90OxGoxvHUqJqEEJ1VU9DkJKaKDGamLAbWAecC\nS4FvANhePeVrn0yz7e/upe+2P9bxPYuNp08kPQZcavsZSccD99peVjquGJ2k+23/4oH+Pkvp7dXl\nS3u93jLja98D3A/81wTfs+R4+mSfOxxJucPpr+cl3ceegmaxpI/D7AuaVO4BgKSv257EfjIxotzh\nLByS3nug/2b7C7OMJZV7tB6XdCXN0vc5ANv/UTakQ0bucBaIWSfwg0nlHgBIenC/T83loOaI/kpy\nj90kHQO8Fdhs+9nS8UTE+DIVMgCQ9BvAQ8Bq4GFJVxUOKSI6SHKP1vXAubYvA84Gfq9wPBHRQZJ7\ntF6x/UMA28/TbNEbET2V2TLR2izpz4D1wHJgU+F4IqKDVO7RuhrYTLMr4SbgmrLhREQXSe7R+jHg\nyzQnMB0DnFA2nIjoIsk9Wn8NnAP8KbALuKNsOBHRRZJ7tBYD9wIn2f4T4HWF44mIDpLco3U4zfTH\nRyWdRnNwRET0VJJ7tG6g6bPfClxM5rlH9Fq2H4jdJC0FDgMWASfY3lA4pIgYU+a5BwCS1tCcHLOE\npv++CbigaFARMba0ZaJ1JnA6cB/w02SFakSvJblHa7vtOWBJdoSM6L/03AOAwVFgzwHHAycDp9o+\nv2xUETGuJPfYTdKP07Rjfhl4xPa2wiFFxJjSlgkAJJ0I3A78HfBm4JSiAUVEJ0nu0boD+CzNVMj1\nwF+UDSciukhyj9aRth+gOTvVZLZMRK8luUdrp6RLgddJuoAk94heS3KP1gdo9nQ/lmYrguvKhhMR\nXWSFarQ+ZPvK0kFExGSkco/WaZKOKh1ERExG5rkHAJKeBE4EngXmaB6s5jSmiJ5Kco+DkvTrtr9c\nOo6IGE3aMjGf7Ose0UNJ7jGfRaUDiIjRJbnHfNK3i+ihJPeIiAUoyT3mk7ZMRA9lEVPsJukM4CeB\nf7f93cGnbysYUkSMKVMhAwBJH2GwjztwPnC37T8vG1VEjCttmWi9E/hZ2x8ClgPZiiCix5Lco7UN\nWDx4fTjw/YKxRERH6bkf4iRtoJnu+Cbge5L+FTgN2F40sIjoJMk90n6JWIDyQDUAkLSMJtEf0X7O\n9gfLRRQRXaRyj9YXgE8A/1M6kIjoLsk9Wt+z/fnSQUTEZKQtEwBIei/NPPdvt5+z/bFyEUVEF6nc\no/U7wN8A/1s6kIjoLsk9Wtttf6J0EBExGWnLBACSvgi8APwLg21+bd9RNKiIGFsq92htHHx88+Bj\nfutH9Fgq99hN0lLgMJptfk+wvaFwSBExplTuAYCkNcCFwBLgSGAzcEHRoCJibNk4LFpnAqcD99Hs\nLbOzbDgR0UWSe7S2254Dlth+tnQwEdFNeu4BgKSPA88BxwMnA6faPr9sVBExriT32E3S64EXGZzI\nZHtb4ZAiYkxJ7vGaJP2q7a+WjiMixpOeexzI20oHEBHjS+Ue+5B0HM3D1VdKxxIR40tyDwAkXQSs\nAXYARwHX2P562agiYlxpy0TrZuAdts8Cfga4pXA8EdFBknu0Xra9FcD2FrKIKaLXsv1AtHZIWgms\nB5bTzHmPiJ5K5R6tq4C3ALfSLGJ6X9lwIqKLVO6HOEkn2X6KZmXqZ2h2hJwDjiOHZUf0VpJ7XD/4\n82n27OHeJviLSwUVEd1kKmTsJulY4BRgo+2cpRrRY+m5BwCSrga+CXwY2CDpisIhRUQHSe7Rug44\ny/blwHnA7xeOJyI6SHKP1nZg1+D1i0DaMhE9lgeqhzhJd7Fndsyjkh4GzqFJ8BHRU0nucftrfO6u\nmUcREROVtswhzvY62+uANwAXD16vAo4oG1lEdJHKPVo3ARcNXq8AvkZzWHZE9FAq92jtsv0DgMHH\nlwvHExEdpHKP1iOS7gQ2AMuAxwrHExEdZIVq7CbpMkDAd2x/pXQ8ETG+tGUCAElHA0cCTwNvlLSq\ncEgR0UHaMtFaC3wHOIPmoI4XyoYTEV2kco/WItvXAgYuAY4uHE9EdJDkHq0fSToCWEKzYjV3BTL3\n2gAAAJtJREFUdRE9luQerU8BNwDfAp4CnigbTkR0keosWnPA+2lOX3oJ+FzZcCKii1Tu0foosMz2\n2cCFwM2F44mIDpLco7Xd9jMAtrcBOwrHExEdZBFTACBpLbAYWAecCywFvgFge3W5yCJiHOm5R+tL\ne73eUiyKiJiIVO4REQtQeu4REQtQkntExAKU5B4RsQAluUdELEBJ7hERC9D/AzSJc81AgcieAAAA\nAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x115fa1588>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "feature_importances = pd.Series(model.feature_importances_, index=X_test.columns)\n",
    "feature_importances.sort()\n",
    "feature_importances.plot.bar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#TODO: Try training the model with the most important features and note difference in overal model prediction score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0    307\n",
       "2.0     84\n",
       "0.0    211\n",
       "dtype: int64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#NOTE: Training Set - Show number of Hom Ref, Hom Var, Het Var datapoints the model was trained on\n",
    "ytrain = pd.DataFrame()\n",
    "ytrain['ytrain'] = y_train\n",
    "pd.value_counts(ytrain['ytrain'].values, sort=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='prediction_step'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.,  0.,  1.,  1.,  0.,  0.,  0.,  1.,  0.,  0.,  1.,  1.,  0.,\n",
       "        0.,  1.,  0.,  1.,  0.,  0.,  0.,  0.,  1.,  1.,  1.,  1.,  0.,\n",
       "        1.,  0.,  0.,  1.,  0.,  1.,  0.,  0.,  0.,  1.,  0.,  0.,  0.,\n",
       "        1.,  1.,  1.,  1.,  0.,  0.,  1.,  0.,  2.,  1.,  0.,  1.,  0.,\n",
       "        1.,  0.,  0.,  0.,  0.,  0.,  0.,  1.,  0.,  1.,  0.,  0.,  0.,\n",
       "        0.,  1.,  1.,  1.,  1.,  1.,  0.,  1.,  0.,  1.,  0.,  0.,  0.,\n",
       "        1.,  0.,  1.,  0.,  1.,  1.,  1.,  1.,  1.,  0.,  1.,  1.,  0.,\n",
       "        0.,  0.,  0.,  1.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,  1.,  1.,\n",
       "        0.,  2.,  1.,  0.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,  0.,  0.,\n",
       "        0.,  0.,  1.,  1.,  0.,  0.,  0.,  1.,  1.,  0.,  2.,  0.,  1.,\n",
       "        2.,  0.,  1.,  1.,  0.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,  1.,\n",
       "        1.,  0.,  0.,  1.,  0.,  1.,  0.,  0.,  1.,  1.,  0.,  0.,  0.,\n",
       "        0.,  0.,  0.,  2.,  0.,  0.,  0.,  1.,  1.,  1.,  0.,  0.,  0.,\n",
       "        0.,  2.,  0.,  1.,  0.,  0.,  0.,  0.,  0.,  1.,  0.,  1.,  1.,\n",
       "        0.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,  1.,  0.,  1.,  0.,  0.,\n",
       "        1.,  0.,  1.,  0.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  0.,\n",
       "        0.,  0.,  1.,  0.,  0.,  0.,  0.,  1.,  0.,  0.,  1.,  1.,  0.,\n",
       "        0.,  1.,  1.,  0.,  0.,  1.,  1.,  0.,  0.,  0.,  0.,  1.,  1.,\n",
       "        1.,  0.,  0.,  0.,  1.,  0.,  0.,  0.,  1.,  0.,  0.,  0.,  0.,\n",
       "        0.,  1.,  1.,  1.,  0.,  1.,  0.,  0.,  1.,  0.,  2.,  0.,  1.,\n",
       "        1.,  0.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,  0.,  0.,  1.,  0.,\n",
       "        1.,  0.,  1.,  0.,  2.,  0.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,\n",
       "        0.,  1.,  0.,  0.,  0.,  0.,  1.,  0.,  1.,  1.,  0.,  0.,  1.,\n",
       "        0.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,  1.,  1.,  1.,  1.,  1.,\n",
       "        1.,  0.,  0.,  0.,  0.,  1.,  0.,  0.,  0.,  0.,  1.,  1.,  0.,\n",
       "        1.,  1.,  0.,  0.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,  0.,  0.,\n",
       "        0.,  0.,  1.,  1.,  0.,  1.,  1.,  1.,  0.,  1.,  0.,  0.,  0.,\n",
       "        1.,  0.,  1.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,  0.,  2.,  0.,\n",
       "        1.,  0.])"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred = model.predict(X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Precision score of the training subset: 0.885\n"
     ]
    }
   ],
   "source": [
    "print('Precision score of the training subset: {:.3f}'.format(precision_score(pred, y_test, average='micro'))) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Size</th>\n",
       "      <th>chrom</th>\n",
       "      <th>end</th>\n",
       "      <th>pacbio.GT</th>\n",
       "      <th>pacbio.alt_alnScore_mean</th>\n",
       "      <th>pacbio.alt_alnScore_std</th>\n",
       "      <th>pacbio.alt_count</th>\n",
       "      <th>pacbio.alt_insertSize_mean</th>\n",
       "      <th>pacbio.alt_insertSize_std</th>\n",
       "      <th>pacbio.alt_reason_alignmentScore</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>start</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1145</th>\n",
       "      <td>-68.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>116491119.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11740.318180</td>\n",
       "      <td>4086.922596</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>116491050.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.536232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>-39.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>180249288.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>9652.519231</td>\n",
       "      <td>3764.203228</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>180249248.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>-31.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>36796498.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15468.56410</td>\n",
       "      <td>7121.614675</td>\n",
       "      <td>39.0</td>\n",
       "      <td>9779.333333</td>\n",
       "      <td>4314.483249</td>\n",
       "      <td>39.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36796466.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>-26.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>64569961.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18599.73913</td>\n",
       "      <td>7637.535695</td>\n",
       "      <td>23.0</td>\n",
       "      <td>11771.434780</td>\n",
       "      <td>4430.129528</td>\n",
       "      <td>23.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11860.695650</td>\n",
       "      <td>4224.703740</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>64569934.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>-22.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>28512612.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>11141.822220</td>\n",
       "      <td>4379.624252</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28512589.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Size  chrom          end  pacbio.GT  pacbio.alt_alnScore_mean  \\\n",
       "1145 -68.0    8.0  116491119.0        0.0                   0.00000   \n",
       "991  -39.0    4.0  180249288.0        0.0                   0.00000   \n",
       "478  -31.0    8.0   36796498.0        2.0               15468.56410   \n",
       "161  -26.0   15.0   64569961.0        1.0               18599.73913   \n",
       "1154 -22.0    9.0   28512612.0        0.0                   0.00000   \n",
       "\n",
       "      pacbio.alt_alnScore_std  pacbio.alt_count  pacbio.alt_insertSize_mean  \\\n",
       "1145                 0.000000               0.0                    0.000000   \n",
       "991                  0.000000               0.0                    0.000000   \n",
       "478               7121.614675              39.0                 9779.333333   \n",
       "161               7637.535695              23.0                11771.434780   \n",
       "1154                 0.000000               0.0                    0.000000   \n",
       "\n",
       "      pacbio.alt_insertSize_std  pacbio.alt_reason_alignmentScore  \\\n",
       "1145                   0.000000                               0.0   \n",
       "991                    0.000000                               0.0   \n",
       "478                 4314.483249                              39.0   \n",
       "161                 4430.129528                              23.0   \n",
       "1154                   0.000000                               0.0   \n",
       "\n",
       "          ...        pacbio.ref_insertSize_mean  pacbio.ref_insertSize_std  \\\n",
       "1145      ...                      11740.318180                4086.922596   \n",
       "991       ...                       9652.519231                3764.203228   \n",
       "478       ...                          0.000000                   0.000000   \n",
       "161       ...                      11860.695650                4224.703740   \n",
       "1154      ...                      11141.822220                4379.624252   \n",
       "\n",
       "      pacbio.ref_reason_alignmentScore  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "1145                              22.0       0.0       0.0         0.0   \n",
       "991                               52.0       0.0       0.0         0.0   \n",
       "478                                0.0       0.0       0.0         0.0   \n",
       "161                               23.0       0.0       0.0         0.0   \n",
       "1154                              45.0       0.0       0.0         0.0   \n",
       "\n",
       "      segdup_pct        start  tandemrep_cnt  tandemrep_pct  \n",
       "1145         0.0  116491050.0            2.0       0.536232  \n",
       "991          0.0  180249248.0            0.0       0.000000  \n",
       "478          0.0   36796466.0            1.0       1.000000  \n",
       "161          0.0   64569934.0            0.0       0.000000  \n",
       "1154         0.0   28512589.0            1.0       1.000000  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add original labels and predicted labels back to the original dataframe\n",
    "df_Xtest = pd.DataFrame(X_test)\n",
    "df_Xtest.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "labels = pd.DataFrame(y_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_Xtest['predicted_label'] = pred\n",
    "df_Xtest['GTcons'] = df_train['GTcons']\n",
    "df_Xtest['chrom'] = df_train['chrom']\n",
    "df_Xtest['start'] = df_train['start']\n",
    "df_Xtest['end'] = df_train['end']\n",
    "# df_Xtest['Y_test'] = labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_Xtest['GTcons'].replace(0.0, 'Homozygous_Reference', inplace=True)\n",
    "df_Xtest['GTcons'].replace(1.0, 'Heterozygous_Variant', inplace=True)\n",
    "df_Xtest['GTcons'].replace(2.0, 'Homozygous_Variant', inplace=True)\n",
    "df_Xtest['predicted_label'].replace(0.0, 'Homozygous_Reference', inplace=True)\n",
    "df_Xtest['predicted_label'].replace(1.0, 'Heterozygous_Variant', inplace=True)\n",
    "df_Xtest['predicted_label'].replace(2.0, 'Homozygous_Variant', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Homozygous_Reference    214\n",
       "Homozygous_Variant       40\n",
       "Heterozygous_Variant    112\n",
       "dtype: int64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(df_Xtest['GTcons'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Homozygous_Reference    214\n",
       "Homozygous_Variant        9\n",
       "Heterozygous_Variant    143\n",
       "dtype: int64"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(df_Xtest['predicted_label'].values, sort=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[108   0   4]\n",
      " [  3 211   0]\n",
      " [ 32   3   5]]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "ytest = df_Xtest['GTcons']\n",
    "predict = df_Xtest['predicted_label']\n",
    "print(confusion_matrix(ytest, predict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Predicted</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <td>108</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <td>3</td>\n",
       "      <td>211</td>\n",
       "      <td>0</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <td>32</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>143</td>\n",
       "      <td>214</td>\n",
       "      <td>9</td>\n",
       "      <td>366</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Predicted             Heterozygous_Variant  Homozygous_Reference  \\\n",
       "True                                                               \n",
       "Heterozygous_Variant                   108                     0   \n",
       "Homozygous_Reference                     3                   211   \n",
       "Homozygous_Variant                      32                     3   \n",
       "All                                    143                   214   \n",
       "\n",
       "Predicted             Homozygous_Variant  All  \n",
       "True                                           \n",
       "Heterozygous_Variant                   4  112  \n",
       "Homozygous_Reference                   0  214  \n",
       "Homozygous_Variant                     5   40  \n",
       "All                                    9  366  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(ytest, predict, rownames=['True'], colnames=['Predicted'], margins=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Reminder:** The original training set only inlcuded 154 [homozygous variant](#hom_ref) examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      precision    recall  f1-score   support\n",
      "\n",
      "Heterozygous_Variant       0.76      0.96      0.85       112\n",
      "Homozygous_Reference       0.99      0.99      0.99       214\n",
      "  Homozygous_Variant       0.56      0.12      0.20        40\n",
      "\n",
      "         avg / total       0.87      0.89      0.86       366\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "print(classification_report(ytest, predict))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Predict\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='predict'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description:\n",
    "\n",
    "   * In the [previous section](#machine_learning), a RF model was trained on svanalyzer data.\n",
    "\n",
    "       * The model was trained using [train/test split](#train_test) where 70% of the data was used to train the model and the model performance was determined by predicting labels for the remaining 30% of the data\n",
    " * Reminder: The labels for this training set and the following [prediction step](#prediction_step) are the consensus genotype (GTcons) labels generated from a preliminary R analysis based on reference and alternate read count:\n",
    "           * Homozygous Reference (0)\n",
    "           * Heterozygous Variant (1)\n",
    "           * Homozygous Variant (2)\n",
    "           \n",
    "   * The trained model is used in the following section to predict labels for 5000 randomly selected Deletions [these datapoints were randomly selected from [union_170509_refalt.sort.vcf](ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/NIST_UnionSVs_05092017/)]\n",
    "   \n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Load Data **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X2 = df_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Impute missing values using KNN **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imputing row 1/3504 with 0 missing, elapsed time: 3.051\n",
      "Imputing row 101/3504 with 0 missing, elapsed time: 3.052\n",
      "Imputing row 201/3504 with 0 missing, elapsed time: 3.052\n",
      "Imputing row 301/3504 with 0 missing, elapsed time: 3.052\n",
      "Imputing row 401/3504 with 0 missing, elapsed time: 3.054\n",
      "Imputing row 501/3504 with 0 missing, elapsed time: 3.055\n",
      "Imputing row 601/3504 with 0 missing, elapsed time: 3.057\n",
      "Imputing row 701/3504 with 0 missing, elapsed time: 3.057\n",
      "Imputing row 801/3504 with 0 missing, elapsed time: 3.058\n",
      "Imputing row 901/3504 with 0 missing, elapsed time: 3.058\n",
      "Imputing row 1001/3504 with 0 missing, elapsed time: 3.061\n",
      "Imputing row 1101/3504 with 0 missing, elapsed time: 3.063\n",
      "Imputing row 1201/3504 with 0 missing, elapsed time: 3.063\n",
      "Imputing row 1301/3504 with 0 missing, elapsed time: 3.065\n",
      "Imputing row 1401/3504 with 0 missing, elapsed time: 3.067\n",
      "Imputing row 1501/3504 with 0 missing, elapsed time: 3.068\n",
      "Imputing row 1601/3504 with 0 missing, elapsed time: 3.068\n",
      "Imputing row 1701/3504 with 0 missing, elapsed time: 3.069\n",
      "Imputing row 1801/3504 with 0 missing, elapsed time: 3.069\n",
      "Imputing row 1901/3504 with 0 missing, elapsed time: 3.070\n",
      "Imputing row 2001/3504 with 0 missing, elapsed time: 3.070\n",
      "Imputing row 2101/3504 with 0 missing, elapsed time: 3.071\n",
      "Imputing row 2201/3504 with 0 missing, elapsed time: 3.072\n",
      "Imputing row 2301/3504 with 0 missing, elapsed time: 3.073\n",
      "Imputing row 2401/3504 with 0 missing, elapsed time: 3.073\n",
      "Imputing row 2501/3504 with 0 missing, elapsed time: 3.074\n",
      "Imputing row 2601/3504 with 0 missing, elapsed time: 3.074\n",
      "Imputing row 2701/3504 with 0 missing, elapsed time: 3.076\n",
      "Imputing row 2801/3504 with 0 missing, elapsed time: 3.076\n",
      "Imputing row 2901/3504 with 0 missing, elapsed time: 3.077\n",
      "Imputing row 3001/3504 with 0 missing, elapsed time: 3.077\n",
      "Imputing row 3101/3504 with 0 missing, elapsed time: 3.079\n",
      "Imputing row 3201/3504 with 0 missing, elapsed time: 3.080\n",
      "Imputing row 3301/3504 with 0 missing, elapsed time: 3.080\n",
      "Imputing row 3401/3504 with 0 missing, elapsed time: 3.081\n",
      "Imputing row 3501/3504 with 0 missing, elapsed time: 3.081\n"
     ]
    }
   ],
   "source": [
    "#Convert dataframe to matrix\n",
    "X2=X2.as_matrix()\n",
    "\n",
    "#Imput missing values from three closest observations\n",
    "X2_imputed=KNN(k=3).complete(X2)\n",
    "X2=pd.DataFrame(X2_imputed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4.0</td>\n",
       "      <td>95498964.0</td>\n",
       "      <td>95499280.0</td>\n",
       "      <td>-316.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.117089</td>\n",
       "      <td>...</td>\n",
       "      <td>215.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19.0</td>\n",
       "      <td>14399929.0</td>\n",
       "      <td>14399954.0</td>\n",
       "      <td>-24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>128.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>14106.2</td>\n",
       "      <td>6697.022584</td>\n",
       "      <td>15.0</td>\n",
       "      <td>8671.60</td>\n",
       "      <td>3889.530045</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8.0</td>\n",
       "      <td>76766532.0</td>\n",
       "      <td>76766569.0</td>\n",
       "      <td>-36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>169.0</td>\n",
       "      <td>228.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>17009.9</td>\n",
       "      <td>6059.876145</td>\n",
       "      <td>20.0</td>\n",
       "      <td>10437.35</td>\n",
       "      <td>3673.796705</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom       start         end   Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0    4.0  95498964.0  95499280.0 -316.0       0.0       0.0         0.0   \n",
       "1   19.0  14399929.0  14399954.0  -24.0       0.0       0.0         1.0   \n",
       "2    8.0  76766532.0  76766569.0  -36.0       0.0       0.0         0.0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0         0.0            1.0       0.117089                ...                  \n",
       "1         1.0            1.0       1.000000                ...                  \n",
       "2         0.0            0.0       0.000000                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            215.0   \n",
       "1                                            128.0   \n",
       "2                                            169.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       255.0                             0.0   \n",
       "1                       163.0                             0.0   \n",
       "2                       228.0                             1.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           12.0                       0.0   \n",
       "1                            9.0                   14106.2   \n",
       "2                            7.0                   17009.9   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0                 0.000000               0.0                        0.00   \n",
       "1              6697.022584              15.0                     8671.60   \n",
       "2              6059.876145              20.0                    10437.35   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                   0.000000                               0.0  \n",
       "1                3889.530045                              15.0  \n",
       "2                3673.796705                              20.0  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dftest_header = list(df_test.columns.values)\n",
    "X2.columns = dftest_header\n",
    "X2.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Size</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.amb_reason_alignmentScore_alignmentScore</th>\n",
       "      <th>pacbio.amb_reason_flanking</th>\n",
       "      <th>pacbio.amb_reason_multimapping</th>\n",
       "      <th>pacbio.amb_reason_same_scores</th>\n",
       "      <th>pacbio.ref_alnScore_mean</th>\n",
       "      <th>pacbio.ref_alnScore_std</th>\n",
       "      <th>pacbio.ref_count</th>\n",
       "      <th>pacbio.ref_insertSize_mean</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4.0</td>\n",
       "      <td>95498964.0</td>\n",
       "      <td>95499280.0</td>\n",
       "      <td>-316.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.117089</td>\n",
       "      <td>...</td>\n",
       "      <td>215.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19.0</td>\n",
       "      <td>14399929.0</td>\n",
       "      <td>14399954.0</td>\n",
       "      <td>-24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>128.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>14106.2</td>\n",
       "      <td>6697.022584</td>\n",
       "      <td>15.0</td>\n",
       "      <td>8671.60</td>\n",
       "      <td>3889.530045</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8.0</td>\n",
       "      <td>76766532.0</td>\n",
       "      <td>76766569.0</td>\n",
       "      <td>-36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>169.0</td>\n",
       "      <td>228.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>17009.9</td>\n",
       "      <td>6059.876145</td>\n",
       "      <td>20.0</td>\n",
       "      <td>10437.35</td>\n",
       "      <td>3673.796705</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom       start         end   Size  refN_cnt  refN_pct  segdup_cnt  \\\n",
       "0    4.0  95498964.0  95499280.0 -316.0       0.0       0.0         0.0   \n",
       "1   19.0  14399929.0  14399954.0  -24.0       0.0       0.0         1.0   \n",
       "2    8.0  76766532.0  76766569.0  -36.0       0.0       0.0         0.0   \n",
       "\n",
       "   segdup_pct  tandemrep_cnt  tandemrep_pct                ...                 \\\n",
       "0         0.0            1.0       0.117089                ...                  \n",
       "1         1.0            1.0       1.000000                ...                  \n",
       "2         0.0            0.0       0.000000                ...                  \n",
       "\n",
       "   pacbio.amb_reason_alignmentScore_alignmentScore  \\\n",
       "0                                            215.0   \n",
       "1                                            128.0   \n",
       "2                                            169.0   \n",
       "\n",
       "   pacbio.amb_reason_flanking  pacbio.amb_reason_multimapping  \\\n",
       "0                       255.0                             0.0   \n",
       "1                       163.0                             0.0   \n",
       "2                       228.0                             1.0   \n",
       "\n",
       "   pacbio.amb_reason_same_scores  pacbio.ref_alnScore_mean  \\\n",
       "0                           12.0                       0.0   \n",
       "1                            9.0                   14106.2   \n",
       "2                            7.0                   17009.9   \n",
       "\n",
       "   pacbio.ref_alnScore_std  pacbio.ref_count  pacbio.ref_insertSize_mean  \\\n",
       "0                 0.000000               0.0                        0.00   \n",
       "1              6697.022584              15.0                     8671.60   \n",
       "2              6059.876145              20.0                    10437.35   \n",
       "\n",
       "   pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \n",
       "0                   0.000000                               0.0  \n",
       "1                3889.530045                              15.0  \n",
       "2                3673.796705                              20.0  \n",
       "\n",
       "[3 rows x 32 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X3 = pd.DataFrame()\n",
    "X3 = X2\n",
    "X3.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Order features\n",
    "X5 = X2.reindex_axis(sorted(X2.columns), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 1.,  1.,  1., ...,  1.,  0.,  0.])"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.predict(X5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='prediction_step'></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred = model.predict(X5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pred_prob = model.predict_proba(X5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X5['predicted_label'] = pred\n",
    "X5['chrom'] = df_test_2['chrom']\n",
    "X5['start'] = df_test_2['start']\n",
    "X5['end'] = df_test_2['end']\n",
    "X5['Size'] = df_test_2['Size']\n",
    "X5['GTcons'] = df_test_2['GTcons']\n",
    "X5['GTsupp'] = df_test_2['GTsupp']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6 = pd.concat([X5, pd.DataFrame(pred_prob, columns=['1','2','3'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6.to_csv('PB_pred_prob_GTcons_DEL_SVanalyzer.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Note: Reformat X6 csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6 = pd.read_csv('/Users/lmc2/NIST/Notebooks/SVanalyzer_ML/DEL/results/PB/PB_pred_prob_GTcons_DEL_SVanalyzer.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6.rename(columns={'1': 'Homozygous_Reference'}, inplace=True)\n",
    "X6.rename(columns={'2': 'Heterozygous_Variant'}, inplace=True)\n",
    "X6.rename(columns={'3': 'Homozygous_Variant'}, inplace=True)\n",
    "X6.rename(columns={'predicted_label': 'PB_predicted_label'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>GTcons</th>\n",
       "      <th>GTsupp</th>\n",
       "      <th>Size</th>\n",
       "      <th>chrom</th>\n",
       "      <th>end</th>\n",
       "      <th>pacbio.GT</th>\n",
       "      <th>pacbio.alt_alnScore_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>pacbio.ref_insertSize_std</th>\n",
       "      <th>pacbio.ref_reason_alignmentScore</th>\n",
       "      <th>PB_predicted_label</th>\n",
       "      <th>refN_cnt</th>\n",
       "      <th>refN_pct</th>\n",
       "      <th>segdup_cnt</th>\n",
       "      <th>segdup_pct</th>\n",
       "      <th>start</th>\n",
       "      <th>tandemrep_cnt</th>\n",
       "      <th>tandemrep_pct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>-316</td>\n",
       "      <td>4</td>\n",
       "      <td>95499280</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15367.42</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>95498964</td>\n",
       "      <td>1</td>\n",
       "      <td>0.117089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>-24</td>\n",
       "      <td>19</td>\n",
       "      <td>14399954</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14568.25</td>\n",
       "      <td>...</td>\n",
       "      <td>3889.530045</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14399929</td>\n",
       "      <td>1</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>-36</td>\n",
       "      <td>8</td>\n",
       "      <td>76766569</td>\n",
       "      <td>1.0</td>\n",
       "      <td>17313.00</td>\n",
       "      <td>...</td>\n",
       "      <td>3673.796705</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>76766532</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Homozygous_Reference  Heterozygous_Variant  Homozygous_Variant  GTcons  \\\n",
       "0                   0.0                   1.0                 0.0       2   \n",
       "1                   0.0                   0.9                 0.1       1   \n",
       "2                   0.0                   0.8                 0.2       1   \n",
       "\n",
       "   GTsupp  Size chrom       end  pacbio.GT  pacbio.alt_alnScore_mean  \\\n",
       "0       1  -316     4  95499280        2.0                  15367.42   \n",
       "1       1   -24    19  14399954        1.0                  14568.25   \n",
       "2       2   -36     8  76766569        1.0                  17313.00   \n",
       "\n",
       "       ...        pacbio.ref_insertSize_std  pacbio.ref_reason_alignmentScore  \\\n",
       "0      ...                         0.000000                               0.0   \n",
       "1      ...                      3889.530045                              15.0   \n",
       "2      ...                      3673.796705                              20.0   \n",
       "\n",
       "   PB_predicted_label  refN_cnt  refN_pct  segdup_cnt  segdup_pct     start  \\\n",
       "0                   1         0         0           0         0.0  95498964   \n",
       "1                   1         0         0           1         1.0  14399929   \n",
       "2                   1         0         0           0         0.0  76766532   \n",
       "\n",
       "   tandemrep_cnt  tandemrep_pct  \n",
       "0              1       0.117089  \n",
       "1              1       1.000000  \n",
       "2              0       0.000000  \n",
       "\n",
       "[3 rows x 38 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X6.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6.to_csv('/Volumes/lesleydata/SVanalyzer_ML/Sept122017/Step3_ML/results/PB_final_GTcons_df_DEL_SVanalyzer.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Label Analysis\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Description:\n",
    "  * The [random forest(RF) model](#train_test) was trained on svanalyzer data. The trained model was used to predict consensus GT labels for the 5000 deletions that were randomly selected from the union_refalt vcf. The following is a comparison of model predicted labels [Conesus GT] to consensus genotype generated by the R script for the 5000 randomly selected datapoints from union_refalt.vcf "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 712  197   12]\n",
      " [  73 1772   67]\n",
      " [  12  602   57]]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import confusion_matrix\n",
    "consensus_GT = X6['GTcons']\n",
    "predict = X6['PB_predicted_label']\n",
    "print(confusion_matrix(consensus_GT, predict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X6['GTcons'].replace(0, 'Homozygous_Reference', inplace=True)\n",
    "X6['GTcons'].replace(1, 'Heterozygous_Variant', inplace=True)\n",
    "X6['GTcons'].replace(2, 'Homozygous_Variant', inplace=True)\n",
    "X6['PB_predicted_label'].replace(0.0, 'Homozygous_Reference', inplace=True)\n",
    "X6['PB_predicted_label'].replace(1.0, 'Heterozygous_Variant', inplace=True)\n",
    "X6['PB_predicted_label'].replace(2.0, 'Homozygous_Variant', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Predicted</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <td>1772</td>\n",
       "      <td>73</td>\n",
       "      <td>67</td>\n",
       "      <td>1912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <td>197</td>\n",
       "      <td>712</td>\n",
       "      <td>12</td>\n",
       "      <td>921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <td>602</td>\n",
       "      <td>12</td>\n",
       "      <td>57</td>\n",
       "      <td>671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>2571</td>\n",
       "      <td>797</td>\n",
       "      <td>136</td>\n",
       "      <td>3504</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Predicted             Heterozygous_Variant  Homozygous_Reference  \\\n",
       "True                                                               \n",
       "Heterozygous_Variant                  1772                    73   \n",
       "Homozygous_Reference                   197                   712   \n",
       "Homozygous_Variant                     602                    12   \n",
       "All                                   2571                   797   \n",
       "\n",
       "Predicted             Homozygous_Variant   All  \n",
       "True                                            \n",
       "Heterozygous_Variant                  67  1912  \n",
       "Homozygous_Reference                  12   921  \n",
       "Homozygous_Variant                    57   671  \n",
       "All                                  136  3504  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(consensus_GT, predict, rownames=['True'], colnames=['Predicted'], margins=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Precision score of the prediction subset: 0.725\n"
     ]
    }
   ],
   "source": [
    "print('Precision score of the prediction subset: {:.3f}'.format(precision_score(consensus_GT, predict, average='micro'))) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** High Confidence Label Analysis**\n",
    "* **Reminder:** The labels predicted by the model are the following consensus genotype:\n",
    "    * Homozygous Reference: 0 \n",
    "    * Heterozygous Variant: 1 \n",
    "    * Homozygous Variant: 2 \n",
    "* Here **high confidence labels** are the labels predicted by the model that were also assigned a predict probability of either 0.9 or 1\n",
    "* The following is an analysis of predicted svanalyzer labels with predict probability >0.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "high_conf_labels = X6[(X6['Homozygous_Reference'] == 1) | (X6['Homozygous_Reference'] == 0.9) | (X6['Heterozygous_Variant'] == 1) | (X6['Heterozygous_Variant'] == 0.9) | (X6['Homozygous_Variant'] == 1) | (X6['Homozygous_Variant'] == 0.9)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Predicted</th>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Heterozygous_Variant</th>\n",
       "      <td>513</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Reference</th>\n",
       "      <td>40</td>\n",
       "      <td>617</td>\n",
       "      <td>0</td>\n",
       "      <td>657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Homozygous_Variant</th>\n",
       "      <td>213</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>766</td>\n",
       "      <td>654</td>\n",
       "      <td>1</td>\n",
       "      <td>1421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Predicted             Heterozygous_Variant  Homozygous_Reference  \\\n",
       "True                                                               \n",
       "Heterozygous_Variant                   513                    31   \n",
       "Homozygous_Reference                    40                   617   \n",
       "Homozygous_Variant                     213                     6   \n",
       "All                                    766                   654   \n",
       "\n",
       "Predicted             Homozygous_Variant   All  \n",
       "True                                            \n",
       "Heterozygous_Variant                   1   545  \n",
       "Homozygous_Reference                   0   657  \n",
       "Homozygous_Variant                     0   219  \n",
       "All                                    1  1421  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_GT = high_conf_labels['GTcons']\n",
    "predict = high_conf_labels['PB_predicted_label']\n",
    "pd.crosstab(consensus_GT, predict, rownames=['True'], colnames=['Predicted'], margins=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NOTE:** There are only 2 Homozygous Reference examples. This may stem from the model being trained on 154 Hom Var examples. Next step, choose a training set with more Hom_Var examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      precision    recall  f1-score   support\n",
      "\n",
      "Heterozygous_Variant       0.67      0.94      0.78       545\n",
      "Homozygous_Reference       0.94      0.94      0.94       657\n",
      "  Homozygous_Variant       0.00      0.00      0.00       219\n",
      "\n",
      "         avg / total       0.69      0.80      0.74      1421\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import classification_report\n",
    "print(classification_report(consensus_GT, predict))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "Compare to crowdsourced results\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PB: Crowdsourced (Crowdvariant) Results\n",
    "![Figure1](https://raw.githubusercontent.com/lesleymaraina/NIST/master/Notebooks/SVanalyzer_ML/DEL/images/PB_CrowdVar1.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Figure1](https://raw.githubusercontent.com/lesleymaraina/NIST/master/Notebooks/SVanalyzer_ML/DEL/images/PB_CrowdVar2.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Figure1](https://raw.githubusercontent.com/lesleymaraina/NIST/master/Notebooks/SVanalyzer_ML/DEL/images/PB_CrowdVar3.png)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [NIHFAES]",
   "language": "python",
   "name": "Python [NIHFAES]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
